// ngx-amap/esm5/directives/amap-ellipse/amap-ellipse.directive.js
/**
 * @fileoverview added by tsickle
 * Generated from: directives/amap-ellipse/amap-ellipse.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Input, Directive, NgZone } from '@angular/core';
import { zip } from 'rxjs';
import { AMapPolygon, PolygonOptions } from '../../base/amap-polygon';
import { AmapEllipseService } from './amap-ellipse.service';
import { LoggerService } from '../../shared/logger/logger.service';
import { EventBinderService } from '../../shared/event-binder.service';
import { getOptions, ChangeFilter } from '../../utils';
/** @type {?} */
var TAG = 'amap-ellipse';
/** @type {?} */
var EllipseOptions = tslib_1.__spread(PolygonOptions, ['center', 'radius']);
var AmapEllipseDirective = /** @class */ (function (_super) {
    tslib_1.__extends(AmapEllipseDirective, _super);
    function AmapEllipseDirective(os, binder, logger, ngZone) {
        var _this = _super.call(this, os, binder) || this;
        _this.os = os;
        _this.binder = binder;
        _this.logger = logger;
        _this.ngZone = ngZone;
        _this.inited = false;
        return _this;
    }
    /**
     * @return {?}
     */
    AmapEllipseDirective.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this.os.destroy();
    };
    /**
     * @param {?} changes
     * @return {?}
     */
    AmapEllipseDirective.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        var _this = this;
        /** @type {?} */
        var filter = ChangeFilter.of(changes);
        /** @type {?} */
        var ellipse = this.get();
        if (!this.inited) {
            this.logger.d(TAG, 'initializing ...');
            /** @type {?} */
            var options = this.options || getOptions(this, EllipseOptions);
            this.logger.d(TAG, 'options:', options);
            this.os.create(options).subscribe((/**
             * @param {?} m
             * @return {?}
             */
            function (m) {
                _this.ngZone.run((/**
                 * @return {?}
                 */
                function () { return _this.naReady.emit(m); }));
                _this.logger.d(TAG, 'ellipse is ready.');
            }));
            this.inited = true;
        }
        else {
            zip(filter.has('center'), ellipse).subscribe((/**
             * @param {?} __0
             * @return {?}
             */
            function (_a) {
                var _b = tslib_1.__read(_a, 2), v = _b[0], p = _b[1];
                return p.setCenter(v);
            }));
            zip(filter.has('options'), ellipse).subscribe((/**
             * @param {?} __0
             * @return {?}
             */
            function (_a) {
                var _b = tslib_1.__read(_a, 2), v = _b[0], p = _b[1];
                return p.setOptions(v || {});
            }));
            zip(filter.has('extData'), ellipse).subscribe((/**
             * @param {?} __0
             * @return {?}
             */
            function (_a) {
                var _b = tslib_1.__read(_a, 2), v = _b[0], p = _b[1];
                return p.setExtData(v);
            }));
        }
        zip(filter.has('hidden'), ellipse).subscribe((/**
         * @param {?} __0
         * @return {?}
         */
        function (_a) {
            var _b = tslib_1.__read(_a, 2), v = _b[0], p = _b[1];
            return (v ? p.hide() : p.show());
        }));
        filter.has('editor').subscribe((/**
         * @param {?} v
         * @return {?}
         */
        function (v) { return _this.os.toggleEditor(v); }));
    };
    /**
     * 获取已创建的 AMap.Ellipse 对象
     */
    /**
     * 获取已创建的 AMap.Ellipse 对象
     * @return {?}
     */
    AmapEllipseDirective.prototype.get = /**
     * 获取已创建的 AMap.Ellipse 对象
     * @return {?}
     */
    function () {
        return this.os.get();
    };
    AmapEllipseDirective.decorators = [
        { type: Directive, args: [{
                    selector: 'amap-ellipse',
                    exportAs: 'ellipse',
                    providers: [AmapEllipseService],
                },] }
    ];
    /** @nocollapse */
    AmapEllipseDirective.ctorParameters = function () { return [
        { type: AmapEllipseService },
        { type: EventBinderService },
        { type: LoggerService },
        { type: NgZone }
    ]; };
    AmapEllipseDirective.propDecorators = {
        center: [{ type: Input }],
        radius: [{ type: Input }],
        options: [{ type: Input }]
    };
    return AmapEllipseDirective;
}(AMapPolygon));
export { AmapEllipseDirective };
if (false) {
    /**
     * 椭圆的中心
     * @type {?}
     */
    AmapEllipseDirective.prototype.center;
    /**
     * 椭圆半径
     * @type {?}
     */
    AmapEllipseDirective.prototype.radius;
    /**
     * 额外: 会覆盖其他属性的配置方式
     * @type {?}
     */
    AmapEllipseDirective.prototype.options;
    /**
     * @type {?}
     * @private
     */
    AmapEllipseDirective.prototype.inited;
    /**
     * @type {?}
     * @protected
     */
    AmapEllipseDirective.prototype.os;
    /**
     * @type {?}
     * @protected
     */
    AmapEllipseDirective.prototype.binder;
    /**
     * @type {?}
     * @private
     */
    AmapEllipseDirective.prototype.logger;
    /**
     * @type {?}
     * @private
     */
    AmapEllipseDirective.prototype.ngZone;
}
