// ngx-amap/esm5/directives/amap-polygon/amap-polygon.directive.js
/**
 * @fileoverview added by tsickle
 * Generated from: directives/amap-polygon/amap-polygon.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Directive, NgZone } from '@angular/core';
import { zip } from 'rxjs';
import { AMapPolygon, PolygonOptions } from '../../base/amap-polygon';
import { AmapPolygonService } from './amap-polygon.service';
import { LoggerService } from '../../shared/logger/logger.service';
import { EventBinderService } from '../../shared/event-binder.service';
import { getOptions, ChangeFilter } from '../../utils';
/** @type {?} */
var TAG = 'amap-polygon';
var AmapPolygonDirective = /** @class */ (function (_super) {
    tslib_1.__extends(AmapPolygonDirective, _super);
    function AmapPolygonDirective(os, binder, logger, ngZone) {
        var _this = _super.call(this, os, binder) || this;
        _this.os = os;
        _this.binder = binder;
        _this.logger = logger;
        _this.ngZone = ngZone;
        _this.inited = false;
        return _this;
    }
    /**
     * @return {?}
     */
    AmapPolygonDirective.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this.os.destroy();
    };
    /**
     * @param {?} changes
     * @return {?}
     */
    AmapPolygonDirective.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        var _this = this;
        /** @type {?} */
        var filter = ChangeFilter.of(changes);
        /** @type {?} */
        var polygon = this.get();
        if (!this.inited) {
            this.logger.d(TAG, 'initializing ...');
            /** @type {?} */
            var options = this.options || getOptions(this, PolygonOptions);
            this.logger.d(TAG, 'options:', options);
            this.os.create(options).subscribe((/**
             * @param {?} m
             * @return {?}
             */
            function (m) {
                _this.ngZone.run((/**
                 * @return {?}
                 */
                function () { return _this.naReady.emit(m); }));
                _this.logger.d(TAG, 'polygon is ready.');
            }));
            this.inited = true;
        }
        else {
            zip(filter.has('path'), polygon).subscribe((/**
             * @param {?} __0
             * @return {?}
             */
            function (_a) {
                var _b = tslib_1.__read(_a, 2), v = _b[0], p = _b[1];
                return p.setPath(v);
            }));
            zip(filter.has('options'), polygon).subscribe((/**
             * @param {?} __0
             * @return {?}
             */
            function (_a) {
                var _b = tslib_1.__read(_a, 2), v = _b[0], p = _b[1];
                return p.setOptions(v || {});
            }));
            zip(filter.has('extData'), polygon).subscribe((/**
             * @param {?} __0
             * @return {?}
             */
            function (_a) {
                var _b = tslib_1.__read(_a, 2), v = _b[0], p = _b[1];
                return p.setExtData(v);
            }));
        }
        zip(filter.has('hidden'), polygon).subscribe((/**
         * @param {?} __0
         * @return {?}
         */
        function (_a) {
            var _b = tslib_1.__read(_a, 2), v = _b[0], p = _b[1];
            return (v ? p.hide() : p.show());
        }));
        filter.has('editor').subscribe((/**
         * @param {?} v
         * @return {?}
         */
        function (v) { return _this.os.toggleEditor(v); }));
    };
    /**
     * 获取已创建的 AMap.Polygon 对象
     */
    /**
     * 获取已创建的 AMap.Polygon 对象
     * @return {?}
     */
    AmapPolygonDirective.prototype.get = /**
     * 获取已创建的 AMap.Polygon 对象
     * @return {?}
     */
    function () {
        return this.os.get();
    };
    AmapPolygonDirective.decorators = [
        { type: Directive, args: [{
                    selector: 'amap-polygon',
                    exportAs: 'polygon',
                    providers: [AmapPolygonService],
                },] }
    ];
    /** @nocollapse */
    AmapPolygonDirective.ctorParameters = function () { return [
        { type: AmapPolygonService },
        { type: EventBinderService },
        { type: LoggerService },
        { type: NgZone }
    ]; };
    return AmapPolygonDirective;
}(AMapPolygon));
export { AmapPolygonDirective };
if (false) {
    /**
     * @type {?}
     * @private
     */
    AmapPolygonDirective.prototype.inited;
    /**
     * @type {?}
     * @protected
     */
    AmapPolygonDirective.prototype.os;
    /**
     * @type {?}
     * @protected
     */
    AmapPolygonDirective.prototype.binder;
    /**
     * @type {?}
     * @private
     */
    AmapPolygonDirective.prototype.logger;
    /**
     * @type {?}
     * @private
     */
    AmapPolygonDirective.prototype.ngZone;
}
