// ngx-amap/esm2015/directives/amap-circle-marker/amap-circle-marker.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: directives/amap-circle-marker/amap-circle-marker.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, NgZone } from '@angular/core';
import { ReplaySubject } from 'rxjs';
import { map } from 'rxjs/operators';
import { AMapService } from '../../shared/amap.service';
import { LoggerService } from '../../shared/logger/logger.service';
/** @type {?} */
const TAG = 'AmapCircleMarker';
export class AmapCircleMarkerService {
    /**
     * @param {?} amaps
     * @param {?} logger
     * @param {?} ngZone
     */
    constructor(amaps, logger, ngZone) {
        this.amaps = amaps;
        this.logger = logger;
        this.ngZone = ngZone;
        this.circle$ = new ReplaySubject(1);
    }
    /**
     * 获取圆点标记
     * @return {?}
     */
    get() {
        return this.circle$.asObservable();
    }
    /**
     * 创建圆点标记 AMap.CircleMarker
     * @param {?} options 选项
     * @return {?}
     */
    create(options) {
        return this.amaps.get().pipe(map((/**
         * @param {?} m
         * @return {?}
         */
        m => {
            options.map = m;
            this.circle = this.ngZone.runOutsideAngular((/**
             * @return {?}
             */
            () => new AMap.CircleMarker(options)));
            this.logger.d(TAG, 'new CircleMarker created.');
            this.circle$.next(this.circle);
            this.circle$.complete();
            return this.circle;
        })));
    }
    /**
     * 销毁圆点标记
     * @return {?}
     */
    destroy() {
        this.get().subscribe((/**
         * @param {?} m
         * @return {?}
         */
        m => {
            m.setMap(null);
            this.logger.d(TAG, 'CircleMarker destroyed.');
            this.circle = null;
        }));
    }
}
AmapCircleMarkerService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
AmapCircleMarkerService.ctorParameters = () => [
    { type: AMapService },
    { type: LoggerService },
    { type: NgZone }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    AmapCircleMarkerService.prototype.circle;
    /**
     * @type {?}
     * @private
     */
    AmapCircleMarkerService.prototype.circle$;
    /**
     * @type {?}
     * @private
     */
    AmapCircleMarkerService.prototype.amaps;
    /**
     * @type {?}
     * @private
     */
    AmapCircleMarkerService.prototype.logger;
    /**
     * @type {?}
     * @private
     */
    AmapCircleMarkerService.prototype.ngZone;
}
