// ngx-amap/esm2015/directives/amap-circle-marker/amap-circle-marker.directive.js
/**
 * @fileoverview added by tsickle
 * Generated from: directives/amap-circle-marker/amap-circle-marker.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, NgZone, } from '@angular/core';
import { zip } from 'rxjs';
import { AMapCircle, CircleOptions } from '../../base/amap-circle';
import { AmapCircleMarkerService } from './amap-circle-marker.service';
import { LoggerService } from '../../shared/logger/logger.service';
import { EventBinderService } from '../../shared/event-binder.service';
import { getOptions, ChangeFilter } from '../../utils';
/** @type {?} */
const TAG = 'amap-circle-marker';
export class AmapCircleMarkerDirective extends AMapCircle {
    /**
     * @param {?} os
     * @param {?} binder
     * @param {?} logger
     * @param {?} ngZone
     */
    constructor(os, binder, logger, ngZone) {
        super(os, binder);
        this.os = os;
        this.binder = binder;
        this.logger = logger;
        this.ngZone = ngZone;
        this.inited = false;
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this.os.destroy();
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    ngOnChanges(changes) {
        /** @type {?} */
        const filter = ChangeFilter.of(changes);
        /** @type {?} */
        const circle = this.get();
        if (!this.inited) {
            this.logger.d(TAG, 'initializing ...');
            /** @type {?} */
            const options = this.options || getOptions(this, CircleOptions);
            this.logger.d(TAG, 'options:', options);
            this.os.create(options).subscribe((/**
             * @param {?} m
             * @return {?}
             */
            m => {
                this.ngZone.run((/**
                 * @return {?}
                 */
                () => this.naReady.emit(m)));
                this.logger.d(TAG, 'circle marker is ready.');
            }));
            this.inited = true;
        }
        else {
            zip(filter.has('center'), circle).subscribe((/**
             * @param {?} __0
             * @return {?}
             */
            ([v, p]) => p.setCenter(v)));
            zip(filter.has('options'), circle).subscribe((/**
             * @param {?} __0
             * @return {?}
             */
            ([v, p]) => p.setOptions(v || {})));
            zip(filter.has('radius'), circle).subscribe((/**
             * @param {?} __0
             * @return {?}
             */
            ([v, p]) => p.setRadius(v)));
            zip(filter.has('extData'), circle).subscribe((/**
             * @param {?} __0
             * @return {?}
             */
            ([v, p]) => p.setExtData(v)));
        }
        zip(filter.has('hidden'), circle).subscribe((/**
         * @param {?} __0
         * @return {?}
         */
        ([v, p]) => (v ? p.hide() : p.show())));
    }
    /**
     * 获取已创建的 AMap.CircleMarker 对象
     * @return {?}
     */
    get() {
        return this.os.get();
    }
}
AmapCircleMarkerDirective.decorators = [
    { type: Directive, args: [{
                selector: 'amap-circle-marker',
                exportAs: 'circle-marker',
                providers: [AmapCircleMarkerService],
            },] }
];
/** @nocollapse */
AmapCircleMarkerDirective.ctorParameters = () => [
    { type: AmapCircleMarkerService },
    { type: EventBinderService },
    { type: LoggerService },
    { type: NgZone }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    AmapCircleMarkerDirective.prototype.inited;
    /**
     * @type {?}
     * @protected
     */
    AmapCircleMarkerDirective.prototype.os;
    /**
     * @type {?}
     * @protected
     */
    AmapCircleMarkerDirective.prototype.binder;
    /**
     * @type {?}
     * @private
     */
    AmapCircleMarkerDirective.prototype.logger;
    /**
     * @type {?}
     * @private
     */
    AmapCircleMarkerDirective.prototype.ngZone;
}
