// ngx-amap/esm5/directives/amap-marker-clusterer/amap-marker-clusterer.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: directives/amap-marker-clusterer/amap-marker-clusterer.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Injectable, NgZone } from '@angular/core';
import { ReplaySubject, zip } from 'rxjs';
import { map } from 'rxjs/operators';
import { AMapService } from '../../shared/amap.service';
import { PluginLoaderService } from '../../shared/plugin-loader.service';
import { LoggerService } from '../../shared/logger/logger.service';
export var AMapMarkerClusterer;
(function (AMapMarkerClusterer) {
    /**
     * @record
     */
    function StyleObject() { }
    AMapMarkerClusterer.StyleObject = StyleObject;
    if (false) {
        /** @type {?|undefined} */
        StyleObject.prototype.url;
        /** @type {?|undefined} */
        StyleObject.prototype.size;
        /** @type {?|undefined} */
        StyleObject.prototype.offset;
        /** @type {?|undefined} */
        StyleObject.prototype.imageOffset;
        /** @type {?|undefined} */
        StyleObject.prototype.textColor;
        /** @type {?|undefined} */
        StyleObject.prototype.textSize;
    }
    /**
     * @record
     */
    function Options() { }
    AMapMarkerClusterer.Options = Options;
    if (false) {
        /**
         * 聚合计算时网格的像素大小，默认60
         * @type {?|undefined}
         */
        Options.prototype.gridSize;
        /**
         * 聚合的最小数量。默认值为2，即小于2个点则不能成为一个聚合
         * @type {?|undefined}
         */
        Options.prototype.minClusterSize;
        /**
         * 最大的聚合级别，大于该级别就不进行相应的聚合。默认值为18
         * @type {?|undefined}
         */
        Options.prototype.maxZoom;
        /**
         * 聚合点的图标位置是否是所有聚合内点的中心点。默认为否
         * @type {?|undefined}
         */
        Options.prototype.averageCenter;
        /**
         * 指定聚合后的点标记的图标样式，可缺省，缺省时为默认样式
         * @type {?|undefined}
         */
        Options.prototype.styles;
        /**
         * 该方法用来实现聚合点的自定义绘制
         * @type {?|undefined}
         */
        Options.prototype.renderClusterMarker;
        /**
         * 点击聚合点时，是否散开，默认值为：true
         * @type {?|undefined}
         */
        Options.prototype.zoomOnClick;
    }
})(AMapMarkerClusterer || (AMapMarkerClusterer = {}));
/** @type {?} */
var TAG = 'AmapMarkerCluster';
var AmapMarkerClustererService = /** @class */ (function () {
    function AmapMarkerClustererService(amaps, plugins, logger, ngZone) {
        this.amaps = amaps;
        this.plugins = plugins;
        this.logger = logger;
        this.ngZone = ngZone;
        this.markerClusterer$ = new ReplaySubject(1);
    }
    /**
     * 获取点聚合
     */
    /**
     * 获取点聚合
     * @return {?}
     */
    AmapMarkerClustererService.prototype.get = /**
     * 获取点聚合
     * @return {?}
     */
    function () {
        return this.markerClusterer$.asObservable();
    };
    /**
     * 创建点聚合 AMap.MarkerClusterer
     * @param options 选项
     */
    /**
     * 创建点聚合 AMap.MarkerClusterer
     * @param {?} options 选项
     * @return {?}
     */
    AmapMarkerClustererService.prototype.create = /**
     * 创建点聚合 AMap.MarkerClusterer
     * @param {?} options 选项
     * @return {?}
     */
    function (options) {
        var _this = this;
        return zip(this.amaps.get(), this.plugins.load('AMap.MarkerClusterer')).pipe(map((/**
         * @param {?} __0
         * @return {?}
         */
        function (_a) {
            var _b = tslib_1.__read(_a, 1), m = _b[0];
            _this.markerClusterer = _this.ngZone.runOutsideAngular((/**
             * @return {?}
             */
            function () { return new AMap.MarkerClusterer(m, [], options); }));
            _this.logger.d(TAG, 'new markerClusterer created.');
            _this.markerClusterer$.next(_this.markerClusterer);
            _this.markerClusterer$.complete();
            return _this.markerClusterer;
        })));
    };
    /**
     * 销毁点聚合
     */
    /**
     * 销毁点聚合
     * @return {?}
     */
    AmapMarkerClustererService.prototype.destroy = /**
     * 销毁点聚合
     * @return {?}
     */
    function () {
        var _this = this;
        this.get().subscribe((/**
         * @param {?} m
         * @return {?}
         */
        function (m) {
            m.clearMarkers();
            m.setMap(null);
            _this.logger.d(TAG, 'markerClusterer destroyed.');
            _this.markerClusterer = null;
        }));
    };
    AmapMarkerClustererService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    AmapMarkerClustererService.ctorParameters = function () { return [
        { type: AMapService },
        { type: PluginLoaderService },
        { type: LoggerService },
        { type: NgZone }
    ]; };
    return AmapMarkerClustererService;
}());
export { AmapMarkerClustererService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    AmapMarkerClustererService.prototype.markerClusterer;
    /**
     * @type {?}
     * @private
     */
    AmapMarkerClustererService.prototype.markerClusterer$;
    /**
     * @type {?}
     * @private
     */
    AmapMarkerClustererService.prototype.amaps;
    /**
     * @type {?}
     * @private
     */
    AmapMarkerClustererService.prototype.plugins;
    /**
     * @type {?}
     * @private
     */
    AmapMarkerClustererService.prototype.logger;
    /**
     * @type {?}
     * @private
     */
    AmapMarkerClustererService.prototype.ngZone;
}
