// ngx-amap/esm2015/directives/amap-ellipse/amap-ellipse.directive.js
/**
 * @fileoverview added by tsickle
 * Generated from: directives/amap-ellipse/amap-ellipse.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Input, Directive, NgZone } from '@angular/core';
import { zip } from 'rxjs';
import { AMapPolygon, PolygonOptions } from '../../base/amap-polygon';
import { AmapEllipseService } from './amap-ellipse.service';
import { LoggerService } from '../../shared/logger/logger.service';
import { EventBinderService } from '../../shared/event-binder.service';
import { getOptions, ChangeFilter } from '../../utils';
/** @type {?} */
const TAG = 'amap-ellipse';
/** @type {?} */
const EllipseOptions = [...PolygonOptions, 'center', 'radius'];
export class AmapEllipseDirective extends AMapPolygon {
    /**
     * @param {?} os
     * @param {?} binder
     * @param {?} logger
     * @param {?} ngZone
     */
    constructor(os, binder, logger, ngZone) {
        super(os, binder);
        this.os = os;
        this.binder = binder;
        this.logger = logger;
        this.ngZone = ngZone;
        this.inited = false;
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this.os.destroy();
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    ngOnChanges(changes) {
        /** @type {?} */
        const filter = ChangeFilter.of(changes);
        /** @type {?} */
        const ellipse = this.get();
        if (!this.inited) {
            this.logger.d(TAG, 'initializing ...');
            /** @type {?} */
            const options = this.options || getOptions(this, EllipseOptions);
            this.logger.d(TAG, 'options:', options);
            this.os.create(options).subscribe((/**
             * @param {?} m
             * @return {?}
             */
            m => {
                this.ngZone.run((/**
                 * @return {?}
                 */
                () => this.naReady.emit(m)));
                this.logger.d(TAG, 'ellipse is ready.');
            }));
            this.inited = true;
        }
        else {
            zip(filter.has('center'), ellipse).subscribe((/**
             * @param {?} __0
             * @return {?}
             */
            ([v, p]) => p.setCenter(v)));
            zip(filter.has('options'), ellipse).subscribe((/**
             * @param {?} __0
             * @return {?}
             */
            ([v, p]) => p.setOptions(v || {})));
            zip(filter.has('extData'), ellipse).subscribe((/**
             * @param {?} __0
             * @return {?}
             */
            ([v, p]) => p.setExtData(v)));
        }
        zip(filter.has('hidden'), ellipse).subscribe((/**
         * @param {?} __0
         * @return {?}
         */
        ([v, p]) => (v ? p.hide() : p.show())));
        filter.has('editor').subscribe((/**
         * @param {?} v
         * @return {?}
         */
        v => this.os.toggleEditor(v)));
    }
    /**
     * 获取已创建的 AMap.Ellipse 对象
     * @return {?}
     */
    get() {
        return this.os.get();
    }
}
AmapEllipseDirective.decorators = [
    { type: Directive, args: [{
                selector: 'amap-ellipse',
                exportAs: 'ellipse',
                providers: [AmapEllipseService],
            },] }
];
/** @nocollapse */
AmapEllipseDirective.ctorParameters = () => [
    { type: AmapEllipseService },
    { type: EventBinderService },
    { type: LoggerService },
    { type: NgZone }
];
AmapEllipseDirective.propDecorators = {
    center: [{ type: Input }],
    radius: [{ type: Input }],
    options: [{ type: Input }]
};
if (false) {
    /**
     * 椭圆的中心
     * @type {?}
     */
    AmapEllipseDirective.prototype.center;
    /**
     * 椭圆半径
     * @type {?}
     */
    AmapEllipseDirective.prototype.radius;
    /**
     * 额外: 会覆盖其他属性的配置方式
     * @type {?}
     */
    AmapEllipseDirective.prototype.options;
    /**
     * @type {?}
     * @private
     */
    AmapEllipseDirective.prototype.inited;
    /**
     * @type {?}
     * @protected
     */
    AmapEllipseDirective.prototype.os;
    /**
     * @type {?}
     * @protected
     */
    AmapEllipseDirective.prototype.binder;
    /**
     * @type {?}
     * @private
     */
    AmapEllipseDirective.prototype.logger;
    /**
     * @type {?}
     * @private
     */
    AmapEllipseDirective.prototype.ngZone;
}
