// ngx-amap/esm5/directives/amap-marker/amap-marker.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: directives/amap-marker/amap-marker.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, NgZone } from '@angular/core';
import { ReplaySubject } from 'rxjs';
import { map } from 'rxjs/operators';
import { AMapService } from '../../shared/amap.service';
import { LoggerService } from '../../shared/logger/logger.service';
/** @type {?} */
var TAG = 'AmapMarker';
var AmapMarkerService = /** @class */ (function () {
    function AmapMarkerService(amaps, logger, ngZone) {
        this.amaps = amaps;
        this.logger = logger;
        this.ngZone = ngZone;
        this.marker$ = new ReplaySubject(1);
    }
    /**
     * 获取点标记
     */
    /**
     * 获取点标记
     * @return {?}
     */
    AmapMarkerService.prototype.get = /**
     * 获取点标记
     * @return {?}
     */
    function () {
        return this.marker$.asObservable();
    };
    /**
     * 创建点标记 AMap.Marker
     * @param options 选项
     * @param addToMap 是否直接加进地图
     */
    /**
     * 创建点标记 AMap.Marker
     * @param {?} options 选项
     * @param {?=} addToMap 是否直接加进地图
     * @return {?}
     */
    AmapMarkerService.prototype.create = /**
     * 创建点标记 AMap.Marker
     * @param {?} options 选项
     * @param {?=} addToMap 是否直接加进地图
     * @return {?}
     */
    function (options, addToMap) {
        var _this = this;
        if (addToMap === void 0) { addToMap = true; }
        return this.amaps.get().pipe(map((/**
         * @param {?} m
         * @return {?}
         */
        function (m) {
            if (addToMap) {
                options.map = m;
            }
            _this.marker = _this.ngZone.runOutsideAngular((/**
             * @return {?}
             */
            function () { return new AMap.Marker(options); }));
            _this.logger.d(TAG, 'new marker created.');
            _this.marker$.next(_this.marker);
            _this.marker$.complete();
            return _this.marker;
        })));
    };
    /**
     * 销毁点标记
     */
    /**
     * 销毁点标记
     * @return {?}
     */
    AmapMarkerService.prototype.destroy = /**
     * 销毁点标记
     * @return {?}
     */
    function () {
        var _this = this;
        this.get().subscribe((/**
         * @param {?} m
         * @return {?}
         */
        function (m) {
            m.setMap(null);
            _this.logger.d(TAG, 'marker destroyed.');
            _this.marker = null;
        }));
    };
    AmapMarkerService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    AmapMarkerService.ctorParameters = function () { return [
        { type: AMapService },
        { type: LoggerService },
        { type: NgZone }
    ]; };
    return AmapMarkerService;
}());
export { AmapMarkerService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    AmapMarkerService.prototype.marker;
    /**
     * @type {?}
     * @private
     */
    AmapMarkerService.prototype.marker$;
    /**
     * @type {?}
     * @private
     */
    AmapMarkerService.prototype.amaps;
    /**
     * @type {?}
     * @private
     */
    AmapMarkerService.prototype.logger;
    /**
     * @type {?}
     * @private
     */
    AmapMarkerService.prototype.ngZone;
}
