// ngx-amap/esm2015/ngx-amap.js
/**
 * @fileoverview added by tsickle
 * Generated from: ngx-amap.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * Generated bundle index. Do not edit.
 */
export { NgxAmapModule, NgxAmapComponent, AmapMarkerDirective, AmapPolylineDirective, AmapInfoWindowComponent, AmapTextComponent, AmapPolygonDirective, AmapBezierCurveDirective, AmapRectangleDirective, AmapCircleDirective, AmapCircleMarkerDirective, AmapEllipseDirective, AmapToolBarDirective, AmapMarkerClustererDirective, AmapHeatmapDirective, InputAmapAutocompleteDirective, AmapPluginLoaderService, AmapUILoaderService, AMapLoaderService, AmapAutocompleteService, UISimpleMarkerDirective, UIAwesomeMarkerDirective, EventBinderService } from './public-api';
export { AMapCircle as ɵx } from './base/amap-circle';
export { AMapOverlay as ɵf } from './base/amap-overlay';
export { AMapPathOverlay as ɵp } from './base/amap-path-overlay';
export { AMapPolygon as ɵt } from './base/amap-polygon';
export { AMapShapeOverlay as ɵq } from './base/amap-shape-overlay';
export { AmapInfoWindowService as ɵk } from './components/amap-info-window/amap-info-window.service';
export { AmapTextService as ɵs } from './components/amap-text/amap-text.service';
export { AmapBezierCurveService as ɵv } from './directives/amap-bezier-curve/amap-bezier-curve.service';
export { AmapCircleMarkerService as ɵz } from './directives/amap-circle-marker/amap-circle-marker.service';
export { AmapCircleService as ɵy } from './directives/amap-circle/amap-circle.service';
export { AmapEllipseService as ɵba } from './directives/amap-ellipse/amap-ellipse.service';
export { AmapHeatmapService as ɵbd } from './directives/amap-heatmap/amap-heatmap.service';
export { AmapMarkerClustererService as ɵbc } from './directives/amap-marker-clusterer/amap-marker-clusterer.service';
export { AmapMarkerService as ɵj } from './directives/amap-marker/amap-marker.service';
export { AmapPolygonService as ɵu } from './directives/amap-polygon/amap-polygon.service';
export { AmapPolylineService as ɵr } from './directives/amap-polyline/amap-polyline.service';
export { AmapRectangleService as ɵw } from './directives/amap-rectangle/amap-rectangle.service';
export { AmapToolBarService as ɵbb } from './directives/amap-tool-bar/amap-tool-bar.service';
export { UIAwesomeMarkerService as ɵbg } from './directives/ui-awesome-marker/ui-awesome-marker.service';
export { UISimpleMarker as ɵbe } from './directives/ui-simple-marker/ui-simple-marker';
export { UISimpleMarkerService as ɵbf } from './directives/ui-simple-marker/ui-simple-marker.service';
export { AMapService as ɵa } from './shared/amap.service';
export { IconService as ɵn } from './shared/icon.service';
export { DebugLoggerService as ɵbh } from './shared/logger/debug-logger.service';
export { LoggerService as ɵd } from './shared/logger/logger.service';
export { MarkerLabelService as ɵo } from './shared/marker-label.service';
export { NGX_AMAP_CONFIG as ɵc } from './shared/ngx-amap-config';
export { PixelService as ɵl } from './shared/pixel.service';
export { PluginLoaderService as ɵe } from './shared/plugin-loader.service';
export { SizeService as ɵm } from './shared/size.service';
