// ngx-amap/esm5/directives/ui-simple-marker/ui-simple-marker.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: directives/ui-simple-marker/ui-simple-marker.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Injectable, NgZone } from '@angular/core';
import { ReplaySubject, zip } from 'rxjs';
import { map } from 'rxjs/operators';
import { AMapService } from '../../shared/amap.service';
import { AmapUILoaderService } from '../../shared/amap-ui-loader.service';
import { LoggerService } from '../../shared/logger/logger.service';
export var AMapUISimpleMarker;
(function (AMapUISimpleMarker) {
    /**
     * @record
     * @template ExtraData
     */
    function Options() { }
    AMapUISimpleMarker.Options = Options;
    if (false) {
        /**
         * 样式主题
         * @type {?|undefined}
         */
        Options.prototype.iconTheme;
        /**
         * 背景图标样式
         * @type {?|undefined}
         */
        Options.prototype.iconStyle;
        /**
         * 图标前景文字
         * @type {?|undefined}
         */
        Options.prototype.iconLabel;
        /**
         * 是否显示定位点
         * @type {?|undefined}
         */
        Options.prototype.showPositionPoint;
        /**
         * 内建的Dom容器上附带的class，多个class name用空格分开
         * @type {?|undefined}
         */
        Options.prototype.containerClassNames;
    }
})(AMapUISimpleMarker || (AMapUISimpleMarker = {}));
/** @type {?} */
var TAG = 'UISimpleMarker';
var UISimpleMarkerService = /** @class */ (function () {
    function UISimpleMarkerService(amaps, uiLoader, logger, ngZone) {
        this.amaps = amaps;
        this.uiLoader = uiLoader;
        this.logger = logger;
        this.ngZone = ngZone;
        this.marker$ = new ReplaySubject(1);
        this.ui = (/** @type {?} */ (this.uiLoader.load('overlay/SimpleMarker')));
    }
    /**
     * 获取点标记
     */
    /**
     * 获取点标记
     * @return {?}
     */
    UISimpleMarkerService.prototype.get = /**
     * 获取点标记
     * @return {?}
     */
    function () {
        return this.marker$.asObservable();
    };
    /**
     * 创建点标记 AMap.SimpleMarker
     * @param options 选项
     * @param addToMap 是否直接加进地图
     */
    /**
     * 创建点标记 AMap.SimpleMarker
     * @param {?} options 选项
     * @param {?=} addToMap 是否直接加进地图
     * @return {?}
     */
    UISimpleMarkerService.prototype.create = /**
     * 创建点标记 AMap.SimpleMarker
     * @param {?} options 选项
     * @param {?=} addToMap 是否直接加进地图
     * @return {?}
     */
    function (options, addToMap) {
        var _this = this;
        if (addToMap === void 0) { addToMap = true; }
        return zip(this.ui, this.amaps.get()).pipe(map((/**
         * @param {?} __0
         * @return {?}
         */
        function (_a) {
            var _b = tslib_1.__read(_a, 2), SimpleMarker = _b[0], m = _b[1];
            if (addToMap) {
                options.map = m;
            }
            _this.marker = _this.ngZone.runOutsideAngular((/**
             * @return {?}
             */
            function () { return new SimpleMarker(options); }));
            _this.logger.d(TAG, 'new simple marker created.');
            _this.marker$.next(_this.marker);
            _this.marker$.complete();
            return _this.marker;
        })));
    };
    /**
     * 销毁点标记
     */
    /**
     * 销毁点标记
     * @return {?}
     */
    UISimpleMarkerService.prototype.destroy = /**
     * 销毁点标记
     * @return {?}
     */
    function () {
        var _this = this;
        this.get().subscribe((/**
         * @param {?} m
         * @return {?}
         */
        function (m) {
            m.setMap(null);
            _this.logger.d(TAG, 'simple marker destroyed.');
            _this.marker = null;
        }));
    };
    UISimpleMarkerService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    UISimpleMarkerService.ctorParameters = function () { return [
        { type: AMapService },
        { type: AmapUILoaderService },
        { type: LoggerService },
        { type: NgZone }
    ]; };
    return UISimpleMarkerService;
}());
export { UISimpleMarkerService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    UISimpleMarkerService.prototype.marker;
    /**
     * @type {?}
     * @private
     */
    UISimpleMarkerService.prototype.marker$;
    /**
     * @type {?}
     * @private
     */
    UISimpleMarkerService.prototype.ui;
    /**
     * @type {?}
     * @private
     */
    UISimpleMarkerService.prototype.amaps;
    /**
     * @type {?}
     * @private
     */
    UISimpleMarkerService.prototype.uiLoader;
    /**
     * @type {?}
     * @private
     */
    UISimpleMarkerService.prototype.logger;
    /**
     * @type {?}
     * @private
     */
    UISimpleMarkerService.prototype.ngZone;
}
