// ngx-amap/esm5/components/amap-info-window/amap-info-window.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: components/amap-info-window/amap-info-window.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Injectable, NgZone } from '@angular/core';
import { ReplaySubject, zip } from 'rxjs';
import { map } from 'rxjs/operators';
import { AMapService } from '../../shared/amap.service';
import { LoggerService } from '../../shared/logger/logger.service';
/** @type {?} */
var TAG = 'AmapInfoWindow';
var AmapInfoWindowService = /** @class */ (function () {
    function AmapInfoWindowService(amaps, logger, ngZone) {
        this.amaps = amaps;
        this.logger = logger;
        this.ngZone = ngZone;
        this.infoWindow$ = new ReplaySubject(1);
    }
    /**
     * 获取信息窗体
     */
    /**
     * 获取信息窗体
     * @return {?}
     */
    AmapInfoWindowService.prototype.get = /**
     * 获取信息窗体
     * @return {?}
     */
    function () {
        return this.infoWindow$.asObservable();
    };
    /**
     * 创建 AMap.InfoWindow
     * @param options 选项
     */
    /**
     * 创建 AMap.InfoWindow
     * @param {?} options 选项
     * @return {?}
     */
    AmapInfoWindowService.prototype.create = /**
     * 创建 AMap.InfoWindow
     * @param {?} options 选项
     * @return {?}
     */
    function (options) {
        var _this = this;
        return this.amaps.get().pipe(map((/**
         * @param {?} m
         * @return {?}
         */
        function (m) {
            _this.map = m;
            _this.infoWindow = _this.ngZone.runOutsideAngular((/**
             * @return {?}
             */
            function () { return new AMap.InfoWindow(options); }));
            _this.logger.d(TAG, 'new InfoWindow created.');
            _this.infoWindow$.next(_this.infoWindow);
            _this.infoWindow$.complete();
            return _this.infoWindow;
        })));
    };
    /**
     * 销毁
     */
    /**
     * 销毁
     * @return {?}
     */
    AmapInfoWindowService.prototype.destroy = /**
     * 销毁
     * @return {?}
     */
    function () {
        var _this = this;
        this.get().subscribe((/**
         * @param {?} w
         * @return {?}
         */
        function (w) {
            w.close();
            _this.logger.d(TAG, 'InfoWindow destroyed.');
            _this.infoWindow = null;
            _this.map = null;
        }));
    };
    /**
     * 打开
     */
    /**
     * 打开
     * @param {?=} position
     * @return {?}
     */
    AmapInfoWindowService.prototype.open = /**
     * 打开
     * @param {?=} position
     * @return {?}
     */
    function (position) {
        var _this = this;
        this.get().subscribe((/**
         * @param {?} w
         * @return {?}
         */
        function (w) { return w.open(_this.map, position); }));
    };
    /**
     * 在覆盖物上打开窗体
     */
    /**
     * 在覆盖物上打开窗体
     * @param {?} marker
     * @return {?}
     */
    AmapInfoWindowService.prototype.openOnMark = /**
     * 在覆盖物上打开窗体
     * @param {?} marker
     * @return {?}
     */
    function (marker) {
        var _this = this;
        zip(marker, this.get()).subscribe((/**
         * @param {?} __0
         * @return {?}
         */
        function (_a) {
            var _b = tslib_1.__read(_a, 2), m = _b[0], w = _b[1];
            w.open(_this.map, m.getPosition());
        }));
    };
    /**
     * 关闭
     */
    /**
     * 关闭
     * @return {?}
     */
    AmapInfoWindowService.prototype.close = /**
     * 关闭
     * @return {?}
     */
    function () {
        this.get().subscribe((/**
         * @param {?} w
         * @return {?}
         */
        function (w) { return w.close(); }));
    };
    AmapInfoWindowService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    AmapInfoWindowService.ctorParameters = function () { return [
        { type: AMapService },
        { type: LoggerService },
        { type: NgZone }
    ]; };
    return AmapInfoWindowService;
}());
export { AmapInfoWindowService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    AmapInfoWindowService.prototype.map;
    /**
     * @type {?}
     * @private
     */
    AmapInfoWindowService.prototype.infoWindow;
    /**
     * @type {?}
     * @private
     */
    AmapInfoWindowService.prototype.infoWindow$;
    /**
     * @type {?}
     * @private
     */
    AmapInfoWindowService.prototype.amaps;
    /**
     * @type {?}
     * @private
     */
    AmapInfoWindowService.prototype.logger;
    /**
     * @type {?}
     * @private
     */
    AmapInfoWindowService.prototype.ngZone;
}
