// ngx-amap/esm2015/directives/amap-heatmap/amap-heatmap.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: directives/amap-heatmap/amap-heatmap.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, NgZone } from '@angular/core';
import { ReplaySubject, zip } from 'rxjs';
import { map } from 'rxjs/operators';
import { AMapService } from '../../shared/amap.service';
import { PluginLoaderService } from '../../shared/plugin-loader.service';
import { LoggerService } from '../../shared/logger/logger.service';
/** @type {?} */
const TAG = 'AmapHeatmap';
export var AMapHeatmap;
(function (AMapHeatmap) {
    /**
     * @record
     */
    function Options() { }
    AMapHeatmap.Options = Options;
    if (false) {
        /**
         * 热力图中单个点的半径，默认：30，单位：pixel
         * @type {?|undefined}
         */
        Options.prototype.radius;
        /**
         * 热力图的渐变区间
         * @type {?|undefined}
         */
        Options.prototype.gradient;
        /**
         * 热力图透明度数组，取值范围[0,1]，0表示完全透明，1表示不透明
         * 默认：[0,1]
         * @type {?|undefined}
         */
        Options.prototype.opacity;
        /**
         * 支持的缩放级别范围，取值范围[3-18]
         * 默认：[3,18]
         * @type {?|undefined}
         */
        Options.prototype.zooms;
        /** @type {?|undefined} */
        Options.prototype.rejectMapMask;
        /** @type {?|undefined} */
        Options.prototype.visible;
        /** @type {?|undefined} */
        Options.prototype.radiusUnit;
        /** @type {?|undefined} */
        Options.prototype.blur;
        /** @type {?|undefined} */
        Options.prototype.zIndex;
        /** @type {?|undefined} */
        Options.prototype.renderOnZooming;
        /* Skipping unnamed member:
        ['3d']?: {
              heightScale?: number;
              heightBezier?: number[];
              gridSize?: number;
              drawGridLine?: boolean;
            };*/
    }
    /**
     * @record
     */
    function Data() { }
    AMapHeatmap.Data = Data;
    if (false) {
        /**
         * 经度
         * @type {?}
         */
        Data.prototype.lng;
        /**
         * 维度
         * @type {?}
         */
        Data.prototype.lat;
        /**
         * 权重
         * @type {?}
         */
        Data.prototype.count;
    }
    /**
     * @record
     */
    function DataSet() { }
    AMapHeatmap.DataSet = DataSet;
    if (false) {
        /**
         * 权重的最大值
         * @type {?|undefined}
         */
        DataSet.prototype.max;
        /**
         * 坐标数据集
         * @type {?}
         */
        DataSet.prototype.data;
    }
})(AMapHeatmap || (AMapHeatmap = {}));
export class AmapHeatmapService {
    /**
     * @param {?} amaps
     * @param {?} plugin
     * @param {?} logger
     * @param {?} ngZone
     */
    constructor(amaps, plugin, logger, ngZone) {
        this.amaps = amaps;
        this.plugin = plugin;
        this.logger = logger;
        this.ngZone = ngZone;
        this.heatmap$ = new ReplaySubject(1);
    }
    /**
     * 获取热力图
     * @return {?}
     */
    get() {
        return this.heatmap$.asObservable();
    }
    /**
     * 创建热力图 AMap.Heatmap
     * @param {?} options 选项
     * @return {?}
     */
    create(options) {
        return zip(this.amaps.get(), this.plugin.load('AMap.Heatmap')).pipe(map((/**
         * @param {?} __0
         * @return {?}
         */
        ([m]) => {
            this.heatmap = this.ngZone.runOutsideAngular((/**
             * @return {?}
             */
            () => new AMap.Heatmap(m, options)));
            this.logger.d(TAG, 'new Heatmap created.');
            this.heatmap$.next(this.heatmap);
            this.heatmap$.complete();
            return this.heatmap;
        })));
    }
    /**
     * 销毁热力图
     * @return {?}
     */
    destroy() {
        this.get().subscribe((/**
         * @param {?} heatmap
         * @return {?}
         */
        heatmap => {
            heatmap.setMap(null);
            this.logger.d(TAG, 'Heatmap destroyed.');
            this.heatmap = null;
        }));
    }
}
AmapHeatmapService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
AmapHeatmapService.ctorParameters = () => [
    { type: AMapService },
    { type: PluginLoaderService },
    { type: LoggerService },
    { type: NgZone }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    AmapHeatmapService.prototype.heatmap;
    /**
     * @type {?}
     * @private
     */
    AmapHeatmapService.prototype.heatmap$;
    /**
     * @type {?}
     * @private
     */
    AmapHeatmapService.prototype.amaps;
    /**
     * @type {?}
     * @private
     */
    AmapHeatmapService.prototype.plugin;
    /**
     * @type {?}
     * @private
     */
    AmapHeatmapService.prototype.logger;
    /**
     * @type {?}
     * @private
     */
    AmapHeatmapService.prototype.ngZone;
}
