// ngx-amap/esm2015/directives/amap-heatmap/amap-heatmap.directive.js
/**
 * @fileoverview added by tsickle
 * Generated from: directives/amap-heatmap/amap-heatmap.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Input, Output, EventEmitter, Directive, NgZone, } from '@angular/core';
import { zip } from 'rxjs';
import { AmapHeatmapService } from './amap-heatmap.service';
import { LoggerService } from '../../shared/logger/logger.service';
import { getOptions, ChangeFilter } from '../../utils';
/** @type {?} */
const TAG = 'amap-heatmap';
/** @type {?} */
const HeatmapOptions = ['radius', 'gradient', 'opacity', 'zooms'];
export class AmapHeatmapDirective {
    /**
     * @param {?} os
     * @param {?} logger
     * @param {?} ngZone
     */
    constructor(os, logger, ngZone) {
        this.os = os;
        this.logger = logger;
        this.ngZone = ngZone;
        /**
         * 额外: 是否隐藏
         */
        this.hidden = false;
        // ---- Events ----
        this.naReady = new EventEmitter();
        this.inited = false;
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this.os.destroy();
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    ngOnChanges(changes) {
        /** @type {?} */
        const filter = ChangeFilter.of(changes);
        /** @type {?} */
        const heatmap = this.get();
        if (!this.inited) {
            this.logger.d(TAG, 'initializing ...');
            /** @type {?} */
            const options = this.options || getOptions(this, HeatmapOptions);
            this.logger.d(TAG, 'options:', options);
            this.os.create(options).subscribe((/**
             * @param {?} m
             * @return {?}
             */
            m => {
                this.ngZone.run((/**
                 * @return {?}
                 */
                () => this.naReady.emit(m)));
                this.logger.d(TAG, 'heatmap is ready.');
            }));
            this.inited = true;
        }
        else {
            zip(filter.has('options'), heatmap).subscribe((/**
             * @param {?} __0
             * @return {?}
             */
            ([v, p]) => p.setOptions(v || {})));
        }
        zip(filter.notEmpty('dataset'), heatmap).subscribe((/**
         * @param {?} __0
         * @return {?}
         */
        ([v, p]) => {
            p.setDataSet(v);
        }));
        zip(filter.has('hidden'), heatmap).subscribe((/**
         * @param {?} __0
         * @return {?}
         */
        ([v, p]) => (v ? p.hide() : p.show())));
    }
    /**
     * 获取已创建的 AMap.Heatmap 对象
     * @return {?}
     */
    get() {
        return this.os.get();
    }
}
AmapHeatmapDirective.decorators = [
    { type: Directive, args: [{
                selector: 'amap-heatmap',
                exportAs: 'heatmap',
                providers: [AmapHeatmapService],
            },] }
];
/** @nocollapse */
AmapHeatmapDirective.ctorParameters = () => [
    { type: AmapHeatmapService },
    { type: LoggerService },
    { type: NgZone }
];
AmapHeatmapDirective.propDecorators = {
    radius: [{ type: Input }],
    gradient: [{ type: Input }],
    opacity: [{ type: Input }],
    zooms: [{ type: Input }],
    hidden: [{ type: Input }],
    options: [{ type: Input }],
    dataset: [{ type: Input }],
    naReady: [{ type: Output }]
};
if (false) {
    /**
     * 热力图中单个点的半径，默认：30，单位：pixel
     * @type {?}
     */
    AmapHeatmapDirective.prototype.radius;
    /**
     * 热力图的渐变区间
     * @type {?}
     */
    AmapHeatmapDirective.prototype.gradient;
    /**
     * 热力图透明度数组，取值范围[0,1]，0表示完全透明，1表示不透明
     * 默认：[0,1]
     * @type {?}
     */
    AmapHeatmapDirective.prototype.opacity;
    /**
     * 支持的缩放级别范围，取值范围[3-18]
     * 默认：[3,18]
     * @type {?}
     */
    AmapHeatmapDirective.prototype.zooms;
    /**
     * 额外: 是否隐藏
     * @type {?}
     */
    AmapHeatmapDirective.prototype.hidden;
    /**
     * 额外: 会覆盖其他属性的配置方式
     * @type {?}
     */
    AmapHeatmapDirective.prototype.options;
    /**
     * 额外: 坐标数据集
     * @type {?}
     */
    AmapHeatmapDirective.prototype.dataset;
    /** @type {?} */
    AmapHeatmapDirective.prototype.naReady;
    /**
     * @type {?}
     * @private
     */
    AmapHeatmapDirective.prototype.inited;
    /**
     * @type {?}
     * @protected
     */
    AmapHeatmapDirective.prototype.os;
    /**
     * @type {?}
     * @private
     */
    AmapHeatmapDirective.prototype.logger;
    /**
     * @type {?}
     * @private
     */
    AmapHeatmapDirective.prototype.ngZone;
}
