// ngx-amap/esm5/directives/amap-circle/amap-circle.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: directives/amap-circle/amap-circle.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Injectable, NgZone } from '@angular/core';
import { ReplaySubject, zip } from 'rxjs';
import { map, switchMapTo } from 'rxjs/operators';
import { AMapService } from '../../shared/amap.service';
import { LoggerService } from '../../shared/logger/logger.service';
import { PluginLoaderService } from '../../shared/plugin-loader.service';
/** @type {?} */
var TAG = 'AmapCircle';
var AmapCircleService = /** @class */ (function () {
    function AmapCircleService(amaps, logger, ngZone, plugins) {
        this.amaps = amaps;
        this.logger = logger;
        this.ngZone = ngZone;
        this.plugins = plugins;
        this.circle$ = new ReplaySubject(1);
        this.editor$ = new ReplaySubject(1);
    }
    /**
     * 获取圆形
     */
    /**
     * 获取圆形
     * @return {?}
     */
    AmapCircleService.prototype.get = /**
     * 获取圆形
     * @return {?}
     */
    function () {
        return this.circle$.asObservable();
    };
    /**
     * 创建圆形 AMap.Circle
     * @param options 选项
     */
    /**
     * 创建圆形 AMap.Circle
     * @param {?} options 选项
     * @return {?}
     */
    AmapCircleService.prototype.create = /**
     * 创建圆形 AMap.Circle
     * @param {?} options 选项
     * @return {?}
     */
    function (options) {
        var _this = this;
        return this.amaps.get().pipe(map((/**
         * @param {?} m
         * @return {?}
         */
        function (m) {
            options.map = m;
            _this.circle = _this.ngZone.runOutsideAngular((/**
             * @return {?}
             */
            function () { return new AMap.Circle(options); }));
            _this.logger.d(TAG, 'new Circle created.');
            _this.circle$.next(_this.circle);
            _this.circle$.complete();
            return _this.circle;
        })));
    };
    /**
     * 销毁圆形
     */
    /**
     * 销毁圆形
     * @return {?}
     */
    AmapCircleService.prototype.destroy = /**
     * 销毁圆形
     * @return {?}
     */
    function () {
        var _this = this;
        this.get().subscribe((/**
         * @param {?} m
         * @return {?}
         */
        function (m) {
            m.setMap(null);
            _this.logger.d(TAG, 'Circle destroyed.');
            _this.circle = null;
        }));
    };
    /**
     * @return {?}
     */
    AmapCircleService.prototype.getEditor = /**
     * @return {?}
     */
    function () {
        // plugin is loading:
        return this.editor$.asObservable();
    };
    /**
     * 加载编辑插件
     */
    /**
     * 加载编辑插件
     * @return {?}
     */
    AmapCircleService.prototype.loadEditor = /**
     * 加载编辑插件
     * @return {?}
     */
    function () {
        return this.plugins.load('AMap.CircleEditor');
    };
    /**
     * 创建编辑器
     */
    /**
     * 创建编辑器
     * @return {?}
     */
    AmapCircleService.prototype.createEditor = /**
     * 创建编辑器
     * @return {?}
     */
    function () {
        var _this = this;
        return this.loadEditor().pipe(switchMapTo(zip(this.amaps.get(), this.get())), map((/**
         * @param {?} __0
         * @return {?}
         */
        function (_a) {
            var _b = tslib_1.__read(_a, 2), m = _b[0], p = _b[1];
            _this.editor = _this.ngZone.runOutsideAngular((/**
             * @return {?}
             */
            function () { return new AMap.CircleEditor(m, p); }));
            _this.logger.d(TAG, 'new CircleEditor created.');
            _this.editor$.next(_this.editor);
            _this.editor$.complete();
            return _this.editor;
        })));
    };
    /**
     * 开关编辑器
     * @param v 开关选项
     */
    /**
     * 开关编辑器
     * @param {?} v 开关选项
     * @return {?}
     */
    AmapCircleService.prototype.toggleEditor = /**
     * 开关编辑器
     * @param {?} v 开关选项
     * @return {?}
     */
    function (v) {
        if (v && !this.editor) {
            this.createEditor().subscribe((/**
             * @param {?} editor
             * @return {?}
             */
            function (editor) { return editor.open(); }));
            return;
        }
        if (this.editor) {
            if (v) {
                this.editor.open();
            }
            else {
                this.editor.close();
            }
        }
    };
    AmapCircleService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    AmapCircleService.ctorParameters = function () { return [
        { type: AMapService },
        { type: LoggerService },
        { type: NgZone },
        { type: PluginLoaderService }
    ]; };
    return AmapCircleService;
}());
export { AmapCircleService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    AmapCircleService.prototype.circle;
    /**
     * @type {?}
     * @private
     */
    AmapCircleService.prototype.circle$;
    /**
     * @type {?}
     * @private
     */
    AmapCircleService.prototype.editor;
    /**
     * @type {?}
     * @private
     */
    AmapCircleService.prototype.editor$;
    /**
     * @type {?}
     * @private
     */
    AmapCircleService.prototype.amaps;
    /**
     * @type {?}
     * @private
     */
    AmapCircleService.prototype.logger;
    /**
     * @type {?}
     * @private
     */
    AmapCircleService.prototype.ngZone;
    /**
     * @type {?}
     * @private
     */
    AmapCircleService.prototype.plugins;
}
