// ngx-amap/esm2015/directives/amap-marker/amap-marker.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: directives/amap-marker/amap-marker.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, NgZone } from '@angular/core';
import { ReplaySubject } from 'rxjs';
import { map } from 'rxjs/operators';
import { AMapService } from '../../shared/amap.service';
import { LoggerService } from '../../shared/logger/logger.service';
/** @type {?} */
const TAG = 'AmapMarker';
export class AmapMarkerService {
    /**
     * @param {?} amaps
     * @param {?} logger
     * @param {?} ngZone
     */
    constructor(amaps, logger, ngZone) {
        this.amaps = amaps;
        this.logger = logger;
        this.ngZone = ngZone;
        this.marker$ = new ReplaySubject(1);
    }
    /**
     * 获取点标记
     * @return {?}
     */
    get() {
        return this.marker$.asObservable();
    }
    /**
     * 创建点标记 AMap.Marker
     * @param {?} options 选项
     * @param {?=} addToMap 是否直接加进地图
     * @return {?}
     */
    create(options, addToMap = true) {
        return this.amaps.get().pipe(map((/**
         * @param {?} m
         * @return {?}
         */
        m => {
            if (addToMap) {
                options.map = m;
            }
            this.marker = this.ngZone.runOutsideAngular((/**
             * @return {?}
             */
            () => new AMap.Marker(options)));
            this.logger.d(TAG, 'new marker created.');
            this.marker$.next(this.marker);
            this.marker$.complete();
            return this.marker;
        })));
    }
    /**
     * 销毁点标记
     * @return {?}
     */
    destroy() {
        this.get().subscribe((/**
         * @param {?} m
         * @return {?}
         */
        m => {
            m.setMap(null);
            this.logger.d(TAG, 'marker destroyed.');
            this.marker = null;
        }));
    }
}
AmapMarkerService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
AmapMarkerService.ctorParameters = () => [
    { type: AMapService },
    { type: LoggerService },
    { type: NgZone }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    AmapMarkerService.prototype.marker;
    /**
     * @type {?}
     * @private
     */
    AmapMarkerService.prototype.marker$;
    /**
     * @type {?}
     * @private
     */
    AmapMarkerService.prototype.amaps;
    /**
     * @type {?}
     * @private
     */
    AmapMarkerService.prototype.logger;
    /**
     * @type {?}
     * @private
     */
    AmapMarkerService.prototype.ngZone;
}
