// ngx-amap/esm2015/components/amap-info-window/amap-info-window.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: components/amap-info-window/amap-info-window.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, NgZone } from '@angular/core';
import { ReplaySubject, zip } from 'rxjs';
import { map } from 'rxjs/operators';
import { AMapService } from '../../shared/amap.service';
import { LoggerService } from '../../shared/logger/logger.service';
/** @type {?} */
const TAG = 'AmapInfoWindow';
export class AmapInfoWindowService {
    /**
     * @param {?} amaps
     * @param {?} logger
     * @param {?} ngZone
     */
    constructor(amaps, logger, ngZone) {
        this.amaps = amaps;
        this.logger = logger;
        this.ngZone = ngZone;
        this.infoWindow$ = new ReplaySubject(1);
    }
    /**
     * 获取信息窗体
     * @return {?}
     */
    get() {
        return this.infoWindow$.asObservable();
    }
    /**
     * 创建 AMap.InfoWindow
     * @param {?} options 选项
     * @return {?}
     */
    create(options) {
        return this.amaps.get().pipe(map((/**
         * @param {?} m
         * @return {?}
         */
        m => {
            this.map = m;
            this.infoWindow = this.ngZone.runOutsideAngular((/**
             * @return {?}
             */
            () => new AMap.InfoWindow(options)));
            this.logger.d(TAG, 'new InfoWindow created.');
            this.infoWindow$.next(this.infoWindow);
            this.infoWindow$.complete();
            return this.infoWindow;
        })));
    }
    /**
     * 销毁
     * @return {?}
     */
    destroy() {
        this.get().subscribe((/**
         * @param {?} w
         * @return {?}
         */
        w => {
            w.close();
            this.logger.d(TAG, 'InfoWindow destroyed.');
            this.infoWindow = null;
            this.map = null;
        }));
    }
    /**
     * 打开
     * @param {?=} position
     * @return {?}
     */
    open(position) {
        this.get().subscribe((/**
         * @param {?} w
         * @return {?}
         */
        w => w.open(this.map, position)));
    }
    /**
     * 在覆盖物上打开窗体
     * @param {?} marker
     * @return {?}
     */
    openOnMark(marker) {
        zip(marker, this.get()).subscribe((/**
         * @param {?} __0
         * @return {?}
         */
        ([m, w]) => {
            w.open(this.map, m.getPosition());
        }));
    }
    /**
     * 关闭
     * @return {?}
     */
    close() {
        this.get().subscribe((/**
         * @param {?} w
         * @return {?}
         */
        w => w.close()));
    }
}
AmapInfoWindowService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
AmapInfoWindowService.ctorParameters = () => [
    { type: AMapService },
    { type: LoggerService },
    { type: NgZone }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    AmapInfoWindowService.prototype.map;
    /**
     * @type {?}
     * @private
     */
    AmapInfoWindowService.prototype.infoWindow;
    /**
     * @type {?}
     * @private
     */
    AmapInfoWindowService.prototype.infoWindow$;
    /**
     * @type {?}
     * @private
     */
    AmapInfoWindowService.prototype.amaps;
    /**
     * @type {?}
     * @private
     */
    AmapInfoWindowService.prototype.logger;
    /**
     * @type {?}
     * @private
     */
    AmapInfoWindowService.prototype.ngZone;
}
