// ngx-amap/esm5/base/amap-path-overlay.js
/**
 * @fileoverview added by tsickle
 * Generated from: base/amap-path-overlay.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Input } from '@angular/core';
import { OverlayOptions } from './amap-overlay';
import { AMapShapeOverlay } from './amap-shape-overlay';
/** @type {?} */
export var PathOverlayOptions = tslib_1.__spread(OverlayOptions, [
    'visible',
    'zIndex',
    'strokeColor',
    'strokeOpacity',
    'strokeWeight',
    'strokeStyle',
    'strokeDasharray',
    'lineJoin',
    'lineCap',
]);
/**
 * @template T
 */
var AMapPathOverlay = /** @class */ (function (_super) {
    tslib_1.__extends(AMapPathOverlay, _super);
    function AMapPathOverlay(os, binder) {
        var _this = _super.call(this, os, binder) || this;
        _this.os = os;
        _this.binder = binder;
        return _this;
    }
    AMapPathOverlay.propDecorators = {
        visible: [{ type: Input }],
        zIndex: [{ type: Input }],
        strokeColor: [{ type: Input }],
        strokeOpacity: [{ type: Input }],
        strokeWeight: [{ type: Input }],
        strokeStyle: [{ type: Input }],
        strokeDasharray: [{ type: Input }],
        lineJoin: [{ type: Input }],
        lineCap: [{ type: Input }]
    };
    return AMapPathOverlay;
}(AMapShapeOverlay));
export { AMapPathOverlay };
if (false) {
    /**
     * 是否可见
     * @type {?}
     */
    AMapPathOverlay.prototype.visible;
    /**
     * 覆盖物层级
     * @type {?}
     */
    AMapPathOverlay.prototype.zIndex;
    /**
     * 描边线条颜色
     * @type {?}
     */
    AMapPathOverlay.prototype.strokeColor;
    /**
     * 描边线条透明度
     * @type {?}
     */
    AMapPathOverlay.prototype.strokeOpacity;
    /**
     * 描边宽度
     * @type {?}
     */
    AMapPathOverlay.prototype.strokeWeight;
    /**
     * 描边样式
     * @type {?}
     */
    AMapPathOverlay.prototype.strokeStyle;
    /**
     * 虚线间隔
     * @type {?}
     */
    AMapPathOverlay.prototype.strokeDasharray;
    /**
     * 折线拐点的绘制样式
     * @type {?}
     */
    AMapPathOverlay.prototype.lineJoin;
    /**
     * 折线两端线帽的绘制样式
     * @type {?}
     */
    AMapPathOverlay.prototype.lineCap;
    /**
     * @type {?}
     * @protected
     */
    AMapPathOverlay.prototype.os;
    /**
     * @type {?}
     * @protected
     */
    AMapPathOverlay.prototype.binder;
}
