// ngx-amap/esm5/base/amap-overlay.js
/**
 * @fileoverview added by tsickle
 * Generated from: base/amap-overlay.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Input, Output, EventEmitter } from '@angular/core';
/** @type {?} */
export var OverlayOptions = ['cursor', 'extData', 'bubble', 'clickable', 'draggable'];
/**
 * @template T
 */
var AMapOverlay = /** @class */ (function () {
    function AMapOverlay(os, binder) {
        this.os = os;
        this.binder = binder;
        /** @type {?} */
        var target = os.get();
        this.naTouchStart = this.binder.bindEvent(target, 'touchstart');
        this.naTouchMove = this.binder.bindEvent(target, 'touchmove');
        this.naTouchEnd = this.binder.bindEvent(target, 'touchend');
        this.naClick = this.binder.bindEvent(target, 'click');
        this.naDblClick = this.binder.bindEvent(target, 'dblclick');
        this.naRightClick = this.binder.bindEvent(target, 'rightclick');
        this.naMouseOver = this.binder.bindEvent(target, 'mouseover');
        this.naMouseUp = this.binder.bindEvent(target, 'mouseup');
        this.naMouseOut = this.binder.bindEvent(target, 'mouseout');
        this.naMouseDown = this.binder.bindEvent(target, 'mousedown');
    }
    AMapOverlay.propDecorators = {
        cursor: [{ type: Input }],
        extData: [{ type: Input }],
        bubble: [{ type: Input }],
        clickable: [{ type: Input }],
        draggable: [{ type: Input }],
        naTouchStart: [{ type: Output }],
        naTouchMove: [{ type: Output }],
        naTouchEnd: [{ type: Output }],
        naClick: [{ type: Output }],
        naDblClick: [{ type: Output }],
        naRightClick: [{ type: Output }],
        naMouseOver: [{ type: Output }],
        naMouseUp: [{ type: Output }],
        naMouseOut: [{ type: Output }],
        naMouseDown: [{ type: Output }]
    };
    return AMapOverlay;
}());
export { AMapOverlay };
if (false) {
    /**
     * 鼠标悬停时的鼠标样式
     * @type {?}
     */
    AMapOverlay.prototype.cursor;
    /**
     * 自定义数据
     * @type {?}
     */
    AMapOverlay.prototype.extData;
    /**
     * 事件是否穿透到地图
     * @type {?}
     */
    AMapOverlay.prototype.bubble;
    /**
     * 是否支持点击
     * @type {?}
     */
    AMapOverlay.prototype.clickable;
    /**
     * 是否支持拖拽
     * @type {?}
     */
    AMapOverlay.prototype.draggable;
    /** @type {?} */
    AMapOverlay.prototype.naTouchStart;
    /** @type {?} */
    AMapOverlay.prototype.naTouchMove;
    /** @type {?} */
    AMapOverlay.prototype.naTouchEnd;
    /** @type {?} */
    AMapOverlay.prototype.naClick;
    /** @type {?} */
    AMapOverlay.prototype.naDblClick;
    /** @type {?} */
    AMapOverlay.prototype.naRightClick;
    /** @type {?} */
    AMapOverlay.prototype.naMouseOver;
    /** @type {?} */
    AMapOverlay.prototype.naMouseUp;
    /** @type {?} */
    AMapOverlay.prototype.naMouseOut;
    /** @type {?} */
    AMapOverlay.prototype.naMouseDown;
    /**
     * @type {?}
     * @protected
     */
    AMapOverlay.prototype.os;
    /**
     * @type {?}
     * @protected
     */
    AMapOverlay.prototype.binder;
}
