// ngx-amap/esm5/base/amap-polygon.js
/**
 * @fileoverview added by tsickle
 * Generated from: base/amap-polygon.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Input, Output, EventEmitter } from '@angular/core';
import { AMapPathOverlay, PathOverlayOptions } from './amap-path-overlay';
/** @type {?} */
export var PolygonOptions = tslib_1.__spread(PathOverlayOptions, ['fillColor', 'fillOpacity', 'path']);
/**
 * @template T, E
 */
var AMapPolygon = /** @class */ (function (_super) {
    tslib_1.__extends(AMapPolygon, _super);
    function AMapPolygon(os, binder) {
        var _this = _super.call(this, os, binder) || this;
        _this.os = os;
        _this.binder = binder;
        /**
         * 额外: 是否隐藏
         */
        _this.hidden = false;
        /**
         * 额外: 是否开启编辑器
         */
        _this.editor = false;
        // directive events:
        _this.naReady = new EventEmitter();
        /** @type {?} */
        var editor = _this.os.getEditor();
        _this.naEditorAddNode = _this.binder.bindEvent(editor, 'addnode');
        _this.naEditorRemoveNode = _this.binder.bindEvent(editor, 'removenode');
        _this.naEditorAdjust = _this.binder.bindEvent(editor, 'adjust');
        _this.naEditorEnd = _this.binder.bindEvent(editor, 'end');
        return _this;
    }
    AMapPolygon.propDecorators = {
        path: [{ type: Input }],
        fillColor: [{ type: Input }],
        fillOpacity: [{ type: Input }],
        hidden: [{ type: Input }],
        editor: [{ type: Input }],
        options: [{ type: Input }],
        naReady: [{ type: Output }],
        naEditorAddNode: [{ type: Output }],
        naEditorRemoveNode: [{ type: Output }],
        naEditorAdjust: [{ type: Output }],
        naEditorEnd: [{ type: Output }]
    };
    return AMapPolygon;
}(AMapPathOverlay));
export { AMapPolygon };
if (false) {
    /**
     * 多边形轮廓线的节点坐标数组
     * @type {?}
     */
    AMapPolygon.prototype.path;
    /**
     * 多边形填充颜色
     * @type {?}
     */
    AMapPolygon.prototype.fillColor;
    /**
     * 边形填充透明度
     * @type {?}
     */
    AMapPolygon.prototype.fillOpacity;
    /**
     * 额外: 是否隐藏
     * @type {?}
     */
    AMapPolygon.prototype.hidden;
    /**
     * 额外: 是否开启编辑器
     * @type {?}
     */
    AMapPolygon.prototype.editor;
    /**
     * 额外: 会覆盖其他属性的配置方式
     * @type {?}
     */
    AMapPolygon.prototype.options;
    /** @type {?} */
    AMapPolygon.prototype.naReady;
    /** @type {?} */
    AMapPolygon.prototype.naEditorAddNode;
    /** @type {?} */
    AMapPolygon.prototype.naEditorRemoveNode;
    /** @type {?} */
    AMapPolygon.prototype.naEditorAdjust;
    /** @type {?} */
    AMapPolygon.prototype.naEditorEnd;
    /**
     * @type {?}
     * @protected
     */
    AMapPolygon.prototype.os;
    /**
     * @type {?}
     * @protected
     */
    AMapPolygon.prototype.binder;
}
