// ngx-amap/esm2015/base/amap-shape-overlay.js
/**
 * @fileoverview added by tsickle
 * Generated from: base/amap-shape-overlay.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Output, EventEmitter } from '@angular/core';
import { AMapOverlay } from './amap-overlay';
/**
 * @template T
 */
export class AMapShapeOverlay extends AMapOverlay {
    /**
     * @param {?} os
     * @param {?} binder
     */
    constructor(os, binder) {
        super(os, binder);
        this.os = os;
        this.binder = binder;
        /** @type {?} */
        const target = os.get();
        this.naShow = this.binder.bindEvent(target, 'show');
        this.naHide = this.binder.bindEvent(target, 'hide');
        this.naChange = this.binder.bindEvent(target, 'change');
    }
}
AMapShapeOverlay.propDecorators = {
    naShow: [{ type: Output }],
    naHide: [{ type: Output }],
    naChange: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    AMapShapeOverlay.prototype.naShow;
    /** @type {?} */
    AMapShapeOverlay.prototype.naHide;
    /** @type {?} */
    AMapShapeOverlay.prototype.naChange;
    /**
     * @type {?}
     * @protected
     */
    AMapShapeOverlay.prototype.os;
    /**
     * @type {?}
     * @protected
     */
    AMapShapeOverlay.prototype.binder;
}
