// ngx-amap/esm2015/base/amap-polygon.js
/**
 * @fileoverview added by tsickle
 * Generated from: base/amap-polygon.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Input, Output, EventEmitter } from '@angular/core';
import { AMapPathOverlay, PathOverlayOptions } from './amap-path-overlay';
/** @type {?} */
export const PolygonOptions = [...PathOverlayOptions, 'fillColor', 'fillOpacity', 'path'];
/**
 * @template T, E
 */
export class AMapPolygon extends AMapPathOverlay {
    /**
     * @param {?} os
     * @param {?} binder
     */
    constructor(os, binder) {
        super(os, binder);
        this.os = os;
        this.binder = binder;
        /**
         * 额外: 是否隐藏
         */
        this.hidden = false;
        /**
         * 额外: 是否开启编辑器
         */
        this.editor = false;
        // directive events:
        this.naReady = new EventEmitter();
        /** @type {?} */
        const editor = this.os.getEditor();
        this.naEditorAddNode = this.binder.bindEvent(editor, 'addnode');
        this.naEditorRemoveNode = this.binder.bindEvent(editor, 'removenode');
        this.naEditorAdjust = this.binder.bindEvent(editor, 'adjust');
        this.naEditorEnd = this.binder.bindEvent(editor, 'end');
    }
}
AMapPolygon.propDecorators = {
    path: [{ type: Input }],
    fillColor: [{ type: Input }],
    fillOpacity: [{ type: Input }],
    hidden: [{ type: Input }],
    editor: [{ type: Input }],
    options: [{ type: Input }],
    naReady: [{ type: Output }],
    naEditorAddNode: [{ type: Output }],
    naEditorRemoveNode: [{ type: Output }],
    naEditorAdjust: [{ type: Output }],
    naEditorEnd: [{ type: Output }]
};
if (false) {
    /**
     * 多边形轮廓线的节点坐标数组
     * @type {?}
     */
    AMapPolygon.prototype.path;
    /**
     * 多边形填充颜色
     * @type {?}
     */
    AMapPolygon.prototype.fillColor;
    /**
     * 边形填充透明度
     * @type {?}
     */
    AMapPolygon.prototype.fillOpacity;
    /**
     * 额外: 是否隐藏
     * @type {?}
     */
    AMapPolygon.prototype.hidden;
    /**
     * 额外: 是否开启编辑器
     * @type {?}
     */
    AMapPolygon.prototype.editor;
    /**
     * 额外: 会覆盖其他属性的配置方式
     * @type {?}
     */
    AMapPolygon.prototype.options;
    /** @type {?} */
    AMapPolygon.prototype.naReady;
    /** @type {?} */
    AMapPolygon.prototype.naEditorAddNode;
    /** @type {?} */
    AMapPolygon.prototype.naEditorRemoveNode;
    /** @type {?} */
    AMapPolygon.prototype.naEditorAdjust;
    /** @type {?} */
    AMapPolygon.prototype.naEditorEnd;
    /**
     * @type {?}
     * @protected
     */
    AMapPolygon.prototype.os;
    /**
     * @type {?}
     * @protected
     */
    AMapPolygon.prototype.binder;
}
