// ngx-amap/esm2015/base/amap-circle.js
/**
 * @fileoverview added by tsickle
 * Generated from: base/amap-circle.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Input, Output, EventEmitter } from '@angular/core';
import { AMapShapeOverlay } from './amap-shape-overlay';
/** @type {?} */
export const CircleOptions = [
    'zIndex',
    'center',
    'bubble',
    'cursor',
    'radius',
    'strokeColor',
    'strokeOpacity',
    'strokeWeight',
    'fillColor',
    'fillOpacity',
    'strokeStyle',
    'extData',
    'strokeDasharray',
];
/**
 * @template T
 */
export class AMapCircle extends AMapShapeOverlay {
    /**
     * @param {?} os
     * @param {?} binder
     */
    constructor(os, binder) {
        super(os, binder);
        this.os = os;
        this.binder = binder;
        /**
         * 额外: 是否隐藏
         */
        this.hidden = false;
        /**
         * 额外: 是否开启编辑器
         */
        this.editor = false;
        // directive events:
        this.naReady = new EventEmitter();
    }
}
AMapCircle.propDecorators = {
    zIndex: [{ type: Input }],
    center: [{ type: Input }],
    radius: [{ type: Input }],
    strokeColor: [{ type: Input }],
    strokeOpacity: [{ type: Input }],
    strokeWeight: [{ type: Input }],
    fillColor: [{ type: Input }],
    fillOpacity: [{ type: Input }],
    strokeStyle: [{ type: Input }],
    strokeDasharray: [{ type: Input }],
    hidden: [{ type: Input }],
    editor: [{ type: Input }],
    options: [{ type: Input }],
    naReady: [{ type: Output }]
};
if (false) {
    /**
     * 层叠顺序，默认zIndex:10
     * @type {?}
     */
    AMapCircle.prototype.zIndex;
    /**
     * 圆心位置
     * @type {?}
     */
    AMapCircle.prototype.center;
    /**
     * 圆半径，单位:米
     * @type {?}
     */
    AMapCircle.prototype.radius;
    /**
     * 线条颜色，使用16进制颜色代码赋值。默认值为#006600
     * @type {?}
     */
    AMapCircle.prototype.strokeColor;
    /**
     * 轮廓线透明度，取值范围[0,1]，0表示完全透明，1表示不透明。默认为0.9
     * @type {?}
     */
    AMapCircle.prototype.strokeOpacity;
    /**
     * 轮廓线宽度
     * @type {?}
     */
    AMapCircle.prototype.strokeWeight;
    /**
     * 圆形填充颜色,使用16进制颜色代码赋值。默认值为#006600
     * @type {?}
     */
    AMapCircle.prototype.fillColor;
    /**
     * 圆形填充透明度，取值范围[0,1]，0表示完全透明，1表示不透明。默认为0.9
     * @type {?}
     */
    AMapCircle.prototype.fillOpacity;
    /**
     * 轮廓线样式，实线:solid，虚线:dashed
     * @type {?}
     */
    AMapCircle.prototype.strokeStyle;
    /**
     * 勾勒形状轮廓的虚线和间隙的样式
     * @type {?}
     */
    AMapCircle.prototype.strokeDasharray;
    /**
     * 额外: 是否隐藏
     * @type {?}
     */
    AMapCircle.prototype.hidden;
    /**
     * 额外: 是否开启编辑器
     * @type {?}
     */
    AMapCircle.prototype.editor;
    /**
     * 额外: 会覆盖其他属性的配置方式
     * @type {?}
     */
    AMapCircle.prototype.options;
    /** @type {?} */
    AMapCircle.prototype.naReady;
    /**
     * @type {?}
     * @protected
     */
    AMapCircle.prototype.os;
    /**
     * @type {?}
     * @protected
     */
    AMapCircle.prototype.binder;
}
