// ngx-amap/esm5/directives/ui-awesome-marker/ui-awesome-marker.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: directives/ui-awesome-marker/ui-awesome-marker.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Injectable, NgZone } from '@angular/core';
import { ReplaySubject, zip } from 'rxjs';
import { map } from 'rxjs/operators';
import { AMapService } from '../../shared/amap.service';
import { AmapUILoaderService } from '../../shared/amap-ui-loader.service';
import { LoggerService } from '../../shared/logger/logger.service';
export var AMapUIAwesomeMarker;
(function (AMapUIAwesomeMarker) {
    /**
     * @record
     * @template ExtraData
     */
    function Options() { }
    AMapUIAwesomeMarker.Options = Options;
    if (false) {
        /**
         * icon 的名称，可用的 icons 参见 Font Awesome 官网
         * @type {?|undefined}
         */
        Options.prototype.awesomeIcon;
        /**
         * 返回字体节点上的 classNames
         * @type {?|undefined}
         */
        Options.prototype.getClassnamesOfAwesomeIcon;
    }
})(AMapUIAwesomeMarker || (AMapUIAwesomeMarker = {}));
/** @type {?} */
var TAG = 'UIAwesomeMarker';
var UIAwesomeMarkerService = /** @class */ (function () {
    function UIAwesomeMarkerService(amaps, uiLoader, logger, ngZone) {
        this.amaps = amaps;
        this.uiLoader = uiLoader;
        this.logger = logger;
        this.ngZone = ngZone;
        this.marker$ = new ReplaySubject(1);
        this.ui = (/** @type {?} */ (this.uiLoader.load('overlay/AwesomeMarker')));
    }
    /**
     * 获取点标记
     */
    /**
     * 获取点标记
     * @return {?}
     */
    UIAwesomeMarkerService.prototype.get = /**
     * 获取点标记
     * @return {?}
     */
    function () {
        return this.marker$.asObservable();
    };
    /**
     * 创建点标记 AMap.AwesomeMarker
     * @param options 选项
     * @param addToMap 是否直接加进地图
     */
    /**
     * 创建点标记 AMap.AwesomeMarker
     * @param {?} options 选项
     * @param {?=} addToMap 是否直接加进地图
     * @return {?}
     */
    UIAwesomeMarkerService.prototype.create = /**
     * 创建点标记 AMap.AwesomeMarker
     * @param {?} options 选项
     * @param {?=} addToMap 是否直接加进地图
     * @return {?}
     */
    function (options, addToMap) {
        var _this = this;
        if (addToMap === void 0) { addToMap = true; }
        return zip(this.ui, this.amaps.get()).pipe(map((/**
         * @param {?} __0
         * @return {?}
         */
        function (_a) {
            var _b = tslib_1.__read(_a, 2), AwesomeMarker = _b[0], m = _b[1];
            if (addToMap) {
                options.map = m;
            }
            _this.marker = _this.ngZone.runOutsideAngular((/**
             * @return {?}
             */
            function () { return new AwesomeMarker(options); }));
            _this.logger.d(TAG, 'new awesome marker created.');
            _this.marker$.next(_this.marker);
            _this.marker$.complete();
            return _this.marker;
        })));
    };
    /**
     * 销毁点标记
     */
    /**
     * 销毁点标记
     * @return {?}
     */
    UIAwesomeMarkerService.prototype.destroy = /**
     * 销毁点标记
     * @return {?}
     */
    function () {
        var _this = this;
        this.get().subscribe((/**
         * @param {?} m
         * @return {?}
         */
        function (m) {
            m.setMap(null);
            _this.logger.d(TAG, 'awesome marker destroyed.');
            _this.marker = null;
        }));
    };
    UIAwesomeMarkerService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    UIAwesomeMarkerService.ctorParameters = function () { return [
        { type: AMapService },
        { type: AmapUILoaderService },
        { type: LoggerService },
        { type: NgZone }
    ]; };
    return UIAwesomeMarkerService;
}());
export { UIAwesomeMarkerService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    UIAwesomeMarkerService.prototype.marker;
    /**
     * @type {?}
     * @private
     */
    UIAwesomeMarkerService.prototype.marker$;
    /**
     * @type {?}
     * @private
     */
    UIAwesomeMarkerService.prototype.ui;
    /**
     * @type {?}
     * @private
     */
    UIAwesomeMarkerService.prototype.amaps;
    /**
     * @type {?}
     * @private
     */
    UIAwesomeMarkerService.prototype.uiLoader;
    /**
     * @type {?}
     * @private
     */
    UIAwesomeMarkerService.prototype.logger;
    /**
     * @type {?}
     * @private
     */
    UIAwesomeMarkerService.prototype.ngZone;
}
