// ngx-amap/esm5/directives/ui-awesome-marker/ui-awesome-marker.directive.js
/**
 * @fileoverview added by tsickle
 * Generated from: directives/ui-awesome-marker/ui-awesome-marker.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Directive, Input, NgZone } from '@angular/core';
import { zip } from 'rxjs';
import { UIAwesomeMarkerService } from './ui-awesome-marker.service';
import { UISimpleMarker, SimpleMarkerOptions } from '../ui-simple-marker/ui-simple-marker';
import { getOptions, ChangeFilter } from '../../utils';
import { EventBinderService } from '../../shared/event-binder.service';
import { AMapService } from '../../shared/amap.service';
import { PixelService } from '../../shared/pixel.service';
import { IconService } from '../../shared/icon.service';
import { MarkerLabelService } from '../../shared/marker-label.service';
import { LoggerService } from '../../shared/logger/logger.service';
/** @type {?} */
var AwesomeMarkerOptions = tslib_1.__spread(SimpleMarkerOptions, ['awesomeIcon', 'getClassnamesOfAwesomeIcon']);
var UIAwesomeMarkerDirective = /** @class */ (function (_super) {
    tslib_1.__extends(UIAwesomeMarkerDirective, _super);
    function UIAwesomeMarkerDirective(os, binder, amaps, pixels, icons, mlabels, logger, ngZone) {
        var _this = _super.call(this, os, binder, amaps, pixels, icons, mlabels, logger, ngZone) || this;
        _this.os = os;
        _this.binder = binder;
        _this.amaps = amaps;
        _this.pixels = pixels;
        _this.icons = icons;
        _this.mlabels = mlabels;
        _this.logger = logger;
        _this.ngZone = ngZone;
        _this.TAG = 'ui-awesome-marker';
        return _this;
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    UIAwesomeMarkerDirective.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        _super.prototype.ngOnChanges.call(this, changes);
        /** @type {?} */
        var filter = ChangeFilter.of(changes);
        /** @type {?} */
        var marker = this.get();
        if (this.inited) {
            zip(filter.has('awesomeIcon'), marker).subscribe((/**
             * @param {?} __0
             * @return {?}
             */
            function (_a) {
                var _b = tslib_1.__read(_a, 2), v = _b[0], m = _b[1];
                return m.setAwesomeIcon(v);
            }));
        }
    };
    /**
     * @return {?}
     */
    UIAwesomeMarkerDirective.prototype.getOptions = /**
     * @return {?}
     */
    function () {
        return getOptions(this, AwesomeMarkerOptions);
    };
    UIAwesomeMarkerDirective.decorators = [
        { type: Directive, args: [{
                    selector: 'ui-awesome-marker',
                    exportAs: 'awesomeMarker',
                    providers: [UIAwesomeMarkerService],
                },] }
    ];
    /** @nocollapse */
    UIAwesomeMarkerDirective.ctorParameters = function () { return [
        { type: UIAwesomeMarkerService },
        { type: EventBinderService },
        { type: AMapService },
        { type: PixelService },
        { type: IconService },
        { type: MarkerLabelService },
        { type: LoggerService },
        { type: NgZone }
    ]; };
    UIAwesomeMarkerDirective.propDecorators = {
        awesomeIcon: [{ type: Input }],
        getClassnamesOfAwesomeIcon: [{ type: Input }]
    };
    return UIAwesomeMarkerDirective;
}(UISimpleMarker));
export { UIAwesomeMarkerDirective };
if (false) {
    /** @type {?} */
    UIAwesomeMarkerDirective.prototype.TAG;
    /**
     * icon 的名称，可用的 icons 参见 Font Awesome 官网
     * @type {?}
     */
    UIAwesomeMarkerDirective.prototype.awesomeIcon;
    /**
     * 返回字体节点上的 classNames
     * @type {?}
     */
    UIAwesomeMarkerDirective.prototype.getClassnamesOfAwesomeIcon;
    /**
     * @type {?}
     * @protected
     */
    UIAwesomeMarkerDirective.prototype.os;
    /**
     * @type {?}
     * @protected
     */
    UIAwesomeMarkerDirective.prototype.binder;
    /**
     * @type {?}
     * @protected
     */
    UIAwesomeMarkerDirective.prototype.amaps;
    /**
     * @type {?}
     * @protected
     */
    UIAwesomeMarkerDirective.prototype.pixels;
    /**
     * @type {?}
     * @protected
     */
    UIAwesomeMarkerDirective.prototype.icons;
    /**
     * @type {?}
     * @protected
     */
    UIAwesomeMarkerDirective.prototype.mlabels;
    /**
     * @type {?}
     * @protected
     */
    UIAwesomeMarkerDirective.prototype.logger;
    /**
     * @type {?}
     * @protected
     */
    UIAwesomeMarkerDirective.prototype.ngZone;
}
