// ngx-amap/esm5/directives/amap-rectangle/amap-rectangle.directive.js
/**
 * @fileoverview added by tsickle
 * Generated from: directives/amap-rectangle/amap-rectangle.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Input, Directive, NgZone } from '@angular/core';
import { zip } from 'rxjs';
import { AMapPolygon, PolygonOptions } from '../../base/amap-polygon';
import { AmapRectangleService } from './amap-rectangle.service';
import { LoggerService } from '../../shared/logger/logger.service';
import { EventBinderService } from '../../shared/event-binder.service';
import { getOptions, ChangeFilter } from '../../utils';
/** @type {?} */
var TAG = 'amap-rectangle';
/** @type {?} */
var RectangleOptions = tslib_1.__spread(PolygonOptions, ['bounds']);
var AmapRectangleDirective = /** @class */ (function (_super) {
    tslib_1.__extends(AmapRectangleDirective, _super);
    function AmapRectangleDirective(os, binder, logger, ngZone) {
        var _this = _super.call(this, os, binder) || this;
        _this.os = os;
        _this.binder = binder;
        _this.logger = logger;
        _this.ngZone = ngZone;
        _this.inited = false;
        return _this;
    }
    /**
     * @return {?}
     */
    AmapRectangleDirective.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this.os.destroy();
    };
    /**
     * @param {?} changes
     * @return {?}
     */
    AmapRectangleDirective.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        var _this = this;
        /** @type {?} */
        var filter = ChangeFilter.of(changes);
        /** @type {?} */
        var rectangle = this.get();
        if (!this.inited) {
            this.logger.d(TAG, 'initializing ...');
            /** @type {?} */
            var options = this.options || getOptions(this, RectangleOptions);
            this.logger.d(TAG, 'options:', options);
            this.os.create(options).subscribe((/**
             * @param {?} m
             * @return {?}
             */
            function (m) {
                _this.ngZone.run((/**
                 * @return {?}
                 */
                function () { return _this.naReady.emit(m); }));
                _this.logger.d(TAG, 'rectangle is ready.');
            }));
            this.inited = true;
        }
        else {
            zip(filter.has('path'), rectangle).subscribe((/**
             * @param {?} __0
             * @return {?}
             */
            function (_a) {
                var _b = tslib_1.__read(_a, 2), v = _b[0], p = _b[1];
                return p.setPath(v);
            }));
            zip(filter.has('options'), rectangle).subscribe((/**
             * @param {?} __0
             * @return {?}
             */
            function (_a) {
                var _b = tslib_1.__read(_a, 2), v = _b[0], p = _b[1];
                return p.setOptions(v || {});
            }));
            zip(filter.has('bounds'), rectangle).subscribe((/**
             * @param {?} __0
             * @return {?}
             */
            function (_a) {
                var _b = tslib_1.__read(_a, 2), v = _b[0], p = _b[1];
                return p.setBounds(v);
            }));
            zip(filter.has('extData'), rectangle).subscribe((/**
             * @param {?} __0
             * @return {?}
             */
            function (_a) {
                var _b = tslib_1.__read(_a, 2), v = _b[0], p = _b[1];
                return p.setExtData(v);
            }));
        }
        zip(filter.has('hidden'), rectangle).subscribe((/**
         * @param {?} __0
         * @return {?}
         */
        function (_a) {
            var _b = tslib_1.__read(_a, 2), v = _b[0], p = _b[1];
            return (v ? p.hide() : p.show());
        }));
        filter.has('editor').subscribe((/**
         * @param {?} v
         * @return {?}
         */
        function (v) { return _this.os.toggleEditor(v); }));
    };
    /**
     * 获取已创建的 AMap.Rectangle 对象
     */
    /**
     * 获取已创建的 AMap.Rectangle 对象
     * @return {?}
     */
    AmapRectangleDirective.prototype.get = /**
     * 获取已创建的 AMap.Rectangle 对象
     * @return {?}
     */
    function () {
        return this.os.get();
    };
    AmapRectangleDirective.decorators = [
        { type: Directive, args: [{
                    selector: 'amap-rectangle',
                    exportAs: 'rectangle',
                    providers: [AmapRectangleService],
                },] }
    ];
    /** @nocollapse */
    AmapRectangleDirective.ctorParameters = function () { return [
        { type: AmapRectangleService },
        { type: EventBinderService },
        { type: LoggerService },
        { type: NgZone }
    ]; };
    AmapRectangleDirective.propDecorators = {
        bounds: [{ type: Input }],
        options: [{ type: Input }]
    };
    return AmapRectangleDirective;
}(AMapPolygon));
export { AmapRectangleDirective };
if (false) {
    /**
     * 矩形的范围
     * @type {?}
     */
    AmapRectangleDirective.prototype.bounds;
    /**
     * 额外: 会覆盖其他属性的配置方式
     * @type {?}
     */
    AmapRectangleDirective.prototype.options;
    /**
     * @type {?}
     * @private
     */
    AmapRectangleDirective.prototype.inited;
    /**
     * @type {?}
     * @protected
     */
    AmapRectangleDirective.prototype.os;
    /**
     * @type {?}
     * @protected
     */
    AmapRectangleDirective.prototype.binder;
    /**
     * @type {?}
     * @private
     */
    AmapRectangleDirective.prototype.logger;
    /**
     * @type {?}
     * @private
     */
    AmapRectangleDirective.prototype.ngZone;
}
