// ngx-amap/esm5/directives/input-amap-autocomplete/input-amap-autocomplete.directive.js
/**
 * @fileoverview added by tsickle
 * Generated from: directives/input-amap-autocomplete/input-amap-autocomplete.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Input, Output, EventEmitter, Directive, ElementRef, NgZone, } from '@angular/core';
import { zip } from 'rxjs';
import { AmapAutocompleteService, } from '../../services/amap-autocomplete/amap-autocomplete.service';
import { LoggerService } from '../../shared/logger/logger.service';
import { EventBinderService } from '../../shared/event-binder.service';
import { getOptions, ChangeFilter } from '../../utils';
/** @type {?} */
var TAG = 'amap-autocomplete';
/** @type {?} */
var AutocompleteOptions = ['type', 'city', 'datatype', 'citylimit'];
var InputAmapAutocompleteDirective = /** @class */ (function () {
    function InputAmapAutocompleteDirective(os, binder, el, logger, ngZone) {
        this.os = os;
        this.binder = binder;
        this.el = el;
        this.logger = logger;
        this.ngZone = ngZone;
        // ---- Events ----
        this.naReady = new EventEmitter();
        this.inited = false;
        /** @type {?} */
        var target = this.os.get();
        this.naComplete = this.binder.bindEvent(target, 'complete');
        this.naSelect = this.binder.bindEvent(target, 'select');
        this.naChoose = this.binder.bindEvent(target, 'choose');
        this.naError = this.binder.bindEvent(target, 'error');
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    InputAmapAutocompleteDirective.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        var _this = this;
        /** @type {?} */
        var filter = ChangeFilter.of(changes);
        /** @type {?} */
        var plugin = this.os.get();
        if (!this.inited) {
            this.logger.d(TAG, 'initializing ...');
            /** @type {?} */
            var options = getOptions(this, AutocompleteOptions);
            options.input = this.el.nativeElement;
            this.logger.d(TAG, 'options:', options);
            this.os.create(options).subscribe((/**
             * @param {?} m
             * @return {?}
             */
            function (m) {
                _this.ngZone.run((/**
                 * @return {?}
                 */
                function () { return _this.naReady.emit(m); }));
                _this.logger.d(TAG, 'autocomplete is ready.');
            }));
            this.inited = true;
        }
        else {
            zip(filter.has('city'), plugin).subscribe((/**
             * @param {?} __0
             * @return {?}
             */
            function (_a) {
                var _b = tslib_1.__read(_a, 2), v = _b[0], p = _b[1];
                return p.setCity(v);
            }));
            zip(filter.has('type'), plugin).subscribe((/**
             * @param {?} __0
             * @return {?}
             */
            function (_a) {
                var _b = tslib_1.__read(_a, 2), v = _b[0], p = _b[1];
                return p.setType(v);
            }));
            zip(filter.has('citylimit'), plugin).subscribe((/**
             * @param {?} __0
             * @return {?}
             */
            function (_a) {
                var _b = tslib_1.__read(_a, 2), v = _b[0], p = _b[1];
                return p.setCityLimit(v);
            }));
        }
    };
    InputAmapAutocompleteDirective.decorators = [
        { type: Directive, args: [{
                    selector: 'input[amapAutocomplete]',
                },] }
    ];
    /** @nocollapse */
    InputAmapAutocompleteDirective.ctorParameters = function () { return [
        { type: AmapAutocompleteService },
        { type: EventBinderService },
        { type: ElementRef },
        { type: LoggerService },
        { type: NgZone }
    ]; };
    InputAmapAutocompleteDirective.propDecorators = {
        type: [{ type: Input }],
        city: [{ type: Input }],
        datatype: [{ type: Input }],
        citylimit: [{ type: Input }],
        naReady: [{ type: Output }],
        naComplete: [{ type: Output }],
        naSelect: [{ type: Output }],
        naChoose: [{ type: Output }],
        naError: [{ type: Output }]
    };
    return InputAmapAutocompleteDirective;
}());
export { InputAmapAutocompleteDirective };
if (false) {
    /**
     * 输入提示时限定POI类型，多个类型用“|”分隔
     * @type {?}
     */
    InputAmapAutocompleteDirective.prototype.type;
    /**
     * 输入提示时限定城市
     * @type {?}
     */
    InputAmapAutocompleteDirective.prototype.city;
    /**
     * 返回的数据类型
     * @type {?}
     */
    InputAmapAutocompleteDirective.prototype.datatype;
    /**
     * 是否强制限制在设置的城市内搜索
     * @type {?}
     */
    InputAmapAutocompleteDirective.prototype.citylimit;
    /** @type {?} */
    InputAmapAutocompleteDirective.prototype.naReady;
    /** @type {?} */
    InputAmapAutocompleteDirective.prototype.naComplete;
    /** @type {?} */
    InputAmapAutocompleteDirective.prototype.naSelect;
    /** @type {?} */
    InputAmapAutocompleteDirective.prototype.naChoose;
    /** @type {?} */
    InputAmapAutocompleteDirective.prototype.naError;
    /**
     * @type {?}
     * @private
     */
    InputAmapAutocompleteDirective.prototype.inited;
    /**
     * @type {?}
     * @protected
     */
    InputAmapAutocompleteDirective.prototype.os;
    /**
     * @type {?}
     * @protected
     */
    InputAmapAutocompleteDirective.prototype.binder;
    /**
     * @type {?}
     * @private
     */
    InputAmapAutocompleteDirective.prototype.el;
    /**
     * @type {?}
     * @private
     */
    InputAmapAutocompleteDirective.prototype.logger;
    /**
     * @type {?}
     * @private
     */
    InputAmapAutocompleteDirective.prototype.ngZone;
}
