// ngx-amap/esm2015/directives/ui-awesome-marker/ui-awesome-marker.directive.js
/**
 * @fileoverview added by tsickle
 * Generated from: directives/ui-awesome-marker/ui-awesome-marker.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, Input, NgZone } from '@angular/core';
import { zip } from 'rxjs';
import { UIAwesomeMarkerService } from './ui-awesome-marker.service';
import { UISimpleMarker, SimpleMarkerOptions } from '../ui-simple-marker/ui-simple-marker';
import { getOptions, ChangeFilter } from '../../utils';
import { EventBinderService } from '../../shared/event-binder.service';
import { AMapService } from '../../shared/amap.service';
import { PixelService } from '../../shared/pixel.service';
import { IconService } from '../../shared/icon.service';
import { MarkerLabelService } from '../../shared/marker-label.service';
import { LoggerService } from '../../shared/logger/logger.service';
/** @type {?} */
const AwesomeMarkerOptions = [...SimpleMarkerOptions, 'awesomeIcon', 'getClassnamesOfAwesomeIcon'];
export class UIAwesomeMarkerDirective extends UISimpleMarker {
    /**
     * @param {?} os
     * @param {?} binder
     * @param {?} amaps
     * @param {?} pixels
     * @param {?} icons
     * @param {?} mlabels
     * @param {?} logger
     * @param {?} ngZone
     */
    constructor(os, binder, amaps, pixels, icons, mlabels, logger, ngZone) {
        super(os, binder, amaps, pixels, icons, mlabels, logger, ngZone);
        this.os = os;
        this.binder = binder;
        this.amaps = amaps;
        this.pixels = pixels;
        this.icons = icons;
        this.mlabels = mlabels;
        this.logger = logger;
        this.ngZone = ngZone;
        this.TAG = 'ui-awesome-marker';
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    ngOnChanges(changes) {
        super.ngOnChanges(changes);
        /** @type {?} */
        const filter = ChangeFilter.of(changes);
        /** @type {?} */
        const marker = this.get();
        if (this.inited) {
            zip(filter.has('awesomeIcon'), marker).subscribe((/**
             * @param {?} __0
             * @return {?}
             */
            ([v, m]) => m.setAwesomeIcon(v)));
        }
    }
    /**
     * @return {?}
     */
    getOptions() {
        return getOptions(this, AwesomeMarkerOptions);
    }
}
UIAwesomeMarkerDirective.decorators = [
    { type: Directive, args: [{
                selector: 'ui-awesome-marker',
                exportAs: 'awesomeMarker',
                providers: [UIAwesomeMarkerService],
            },] }
];
/** @nocollapse */
UIAwesomeMarkerDirective.ctorParameters = () => [
    { type: UIAwesomeMarkerService },
    { type: EventBinderService },
    { type: AMapService },
    { type: PixelService },
    { type: IconService },
    { type: MarkerLabelService },
    { type: LoggerService },
    { type: NgZone }
];
UIAwesomeMarkerDirective.propDecorators = {
    awesomeIcon: [{ type: Input }],
    getClassnamesOfAwesomeIcon: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    UIAwesomeMarkerDirective.prototype.TAG;
    /**
     * icon 的名称，可用的 icons 参见 Font Awesome 官网
     * @type {?}
     */
    UIAwesomeMarkerDirective.prototype.awesomeIcon;
    /**
     * 返回字体节点上的 classNames
     * @type {?}
     */
    UIAwesomeMarkerDirective.prototype.getClassnamesOfAwesomeIcon;
    /**
     * @type {?}
     * @protected
     */
    UIAwesomeMarkerDirective.prototype.os;
    /**
     * @type {?}
     * @protected
     */
    UIAwesomeMarkerDirective.prototype.binder;
    /**
     * @type {?}
     * @protected
     */
    UIAwesomeMarkerDirective.prototype.amaps;
    /**
     * @type {?}
     * @protected
     */
    UIAwesomeMarkerDirective.prototype.pixels;
    /**
     * @type {?}
     * @protected
     */
    UIAwesomeMarkerDirective.prototype.icons;
    /**
     * @type {?}
     * @protected
     */
    UIAwesomeMarkerDirective.prototype.mlabels;
    /**
     * @type {?}
     * @protected
     */
    UIAwesomeMarkerDirective.prototype.logger;
    /**
     * @type {?}
     * @protected
     */
    UIAwesomeMarkerDirective.prototype.ngZone;
}
