// ngx-amap/esm2015/directives/ui-awesome-marker/ui-awesome-marker.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: directives/ui-awesome-marker/ui-awesome-marker.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, NgZone } from '@angular/core';
import { ReplaySubject, zip } from 'rxjs';
import { map } from 'rxjs/operators';
import { AMapService } from '../../shared/amap.service';
import { AmapUILoaderService } from '../../shared/amap-ui-loader.service';
import { LoggerService } from '../../shared/logger/logger.service';
export var AMapUIAwesomeMarker;
(function (AMapUIAwesomeMarker) {
    /**
     * @record
     * @template ExtraData
     */
    function Options() { }
    AMapUIAwesomeMarker.Options = Options;
    if (false) {
        /**
         * icon 的名称，可用的 icons 参见 Font Awesome 官网
         * @type {?|undefined}
         */
        Options.prototype.awesomeIcon;
        /**
         * 返回字体节点上的 classNames
         * @type {?|undefined}
         */
        Options.prototype.getClassnamesOfAwesomeIcon;
    }
})(AMapUIAwesomeMarker || (AMapUIAwesomeMarker = {}));
/** @type {?} */
const TAG = 'UIAwesomeMarker';
export class UIAwesomeMarkerService {
    /**
     * @param {?} amaps
     * @param {?} uiLoader
     * @param {?} logger
     * @param {?} ngZone
     */
    constructor(amaps, uiLoader, logger, ngZone) {
        this.amaps = amaps;
        this.uiLoader = uiLoader;
        this.logger = logger;
        this.ngZone = ngZone;
        this.marker$ = new ReplaySubject(1);
        this.ui = (/** @type {?} */ (this.uiLoader.load('overlay/AwesomeMarker')));
    }
    /**
     * 获取点标记
     * @return {?}
     */
    get() {
        return this.marker$.asObservable();
    }
    /**
     * 创建点标记 AMap.AwesomeMarker
     * @param {?} options 选项
     * @param {?=} addToMap 是否直接加进地图
     * @return {?}
     */
    create(options, addToMap = true) {
        return zip(this.ui, this.amaps.get()).pipe(map((/**
         * @param {?} __0
         * @return {?}
         */
        ([AwesomeMarker, m]) => {
            if (addToMap) {
                options.map = m;
            }
            this.marker = this.ngZone.runOutsideAngular((/**
             * @return {?}
             */
            () => new AwesomeMarker(options)));
            this.logger.d(TAG, 'new awesome marker created.');
            this.marker$.next(this.marker);
            this.marker$.complete();
            return this.marker;
        })));
    }
    /**
     * 销毁点标记
     * @return {?}
     */
    destroy() {
        this.get().subscribe((/**
         * @param {?} m
         * @return {?}
         */
        m => {
            m.setMap(null);
            this.logger.d(TAG, 'awesome marker destroyed.');
            this.marker = null;
        }));
    }
}
UIAwesomeMarkerService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
UIAwesomeMarkerService.ctorParameters = () => [
    { type: AMapService },
    { type: AmapUILoaderService },
    { type: LoggerService },
    { type: NgZone }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    UIAwesomeMarkerService.prototype.marker;
    /**
     * @type {?}
     * @private
     */
    UIAwesomeMarkerService.prototype.marker$;
    /**
     * @type {?}
     * @private
     */
    UIAwesomeMarkerService.prototype.ui;
    /**
     * @type {?}
     * @private
     */
    UIAwesomeMarkerService.prototype.amaps;
    /**
     * @type {?}
     * @private
     */
    UIAwesomeMarkerService.prototype.uiLoader;
    /**
     * @type {?}
     * @private
     */
    UIAwesomeMarkerService.prototype.logger;
    /**
     * @type {?}
     * @private
     */
    UIAwesomeMarkerService.prototype.ngZone;
}
