// ngx-amap/esm2015/shared/amap.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: shared/amap.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, NgZone } from '@angular/core';
import { ReplaySubject } from 'rxjs';
import { map } from 'rxjs/operators';
import { AMapLoaderService } from './amap-loader.service';
import { LoggerService } from './logger/logger.service';
/** @type {?} */
const TAG = 'AMap';
export class AMapService {
    /**
     * @param {?} loader
     * @param {?} logger
     * @param {?} ngZone
     */
    constructor(loader, logger, ngZone) {
        this.loader = loader;
        this.logger = logger;
        this.ngZone = ngZone;
        this.map$ = new ReplaySubject();
    }
    /**
     * 获取地图对象 AMap.Map
     * @return {?}
     */
    get() {
        return this.map$.asObservable();
    }
    /**
     * 创建地图对象 AMap.Map
     * @param {?} container 地图容器的DOM元素
     * @param {?} options 选项
     * @return {?}
     */
    create(container, options) {
        return this.loader.load().pipe(map((/**
         * @return {?}
         */
        () => {
            this.map = this.ngZone.runOutsideAngular((/**
             * @return {?}
             */
            () => new AMap.Map(container, options)));
            this.logger.d(TAG, 'new map created.');
            this.map$.next(this.map);
            this.map$.complete();
            return this.map;
        })));
    }
    /**
     * 销毁地图对象
     * @return {?}
     */
    destroy() {
        this.get().subscribe((/**
         * @param {?} m
         * @return {?}
         */
        m => {
            m.destroy();
            this.logger.d(TAG, 'map destroyed.');
            this.map = null;
        }));
    }
}
AMapService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
AMapService.ctorParameters = () => [
    { type: AMapLoaderService },
    { type: LoggerService },
    { type: NgZone }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    AMapService.prototype.map;
    /**
     * @type {?}
     * @private
     */
    AMapService.prototype.map$;
    /**
     * @type {?}
     * @private
     */
    AMapService.prototype.loader;
    /**
     * @type {?}
     * @private
     */
    AMapService.prototype.logger;
    /**
     * @type {?}
     * @private
     */
    AMapService.prototype.ngZone;
}
