// ngx-amap/esm2015/directives/amap-tool-bar/amap-tool-bar.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: directives/amap-tool-bar/amap-tool-bar.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, NgZone } from '@angular/core';
import { ReplaySubject, zip } from 'rxjs';
import { map } from 'rxjs/operators';
import { AMapService } from '../../shared/amap.service';
import { PluginLoaderService } from '../../shared/plugin-loader.service';
import { LoggerService } from '../../shared/logger/logger.service';
/** @type {?} */
const TAG = 'AmapToolBar';
export var AMapToolBar;
(function (AMapToolBar) {
    /**
     * @record
     */
    function Options() { }
    AMapToolBar.Options = Options;
    if (false) {
        /**
         * 相对于地图容器左上角的偏移量
         * @type {?|undefined}
         */
        Options.prototype.offset;
        /**
         * 控件停靠位置
         * LT:左上角;
         * RT:右上角;
         * LB:左下角;
         * RB:右下角;
         * @type {?|undefined}
         */
        Options.prototype.position;
        /**
         * 标尺键盘是否可见
         * @type {?|undefined}
         */
        Options.prototype.ruler;
        /**
         * 定位失败后，是否开启IP定位
         * @type {?|undefined}
         */
        Options.prototype.noIpLocate;
        /**
         * 是否显示定位按钮
         * @type {?|undefined}
         */
        Options.prototype.locate;
        /**
         * 是否使用精简模式
         * @type {?|undefined}
         */
        Options.prototype.liteStyle;
        /**
         * 方向键盘是否可见
         * @type {?|undefined}
         */
        Options.prototype.direction;
        /**
         * 是否自动定位，即地图初始化加载完成后，是否自动定位的用户所在地，仅在支持HTML5的浏览器中有效
         * @type {?|undefined}
         */
        Options.prototype.autoPosition;
        /**
         * 自定义定位图标，值为Marker对象
         * @type {?|undefined}
         */
        Options.prototype.locationMarker;
        /**
         * 是否使用高德定位sdk用来辅助优化定位效果
         * @type {?|undefined}
         */
        Options.prototype.useNative;
        /** @type {?|undefined} */
        Options.prototype.timeout;
    }
})(AMapToolBar || (AMapToolBar = {}));
export class AmapToolBarService {
    /**
     * @param {?} amaps
     * @param {?} plugin
     * @param {?} logger
     * @param {?} ngZone
     */
    constructor(amaps, plugin, logger, ngZone) {
        this.amaps = amaps;
        this.plugin = plugin;
        this.logger = logger;
        this.ngZone = ngZone;
        this.toolBar$ = new ReplaySubject(1);
    }
    /**
     * 获取工具条
     * @return {?}
     */
    get() {
        return this.toolBar$.asObservable();
    }
    /**
     * 创建工具条 AMap.ToolBar
     * @param {?} options 选项
     * @return {?}
     */
    create(options) {
        return zip(this.amaps.get(), this.plugin.load('AMap.ToolBar')).pipe(map((/**
         * @param {?} __0
         * @return {?}
         */
        ([m]) => {
            this.toolBar = this.ngZone.runOutsideAngular((/**
             * @return {?}
             */
            () => new AMap.ToolBar(options)));
            this.logger.d(TAG, 'new toolBar created.');
            m.addControl(this.toolBar);
            this.toolBar$.next(this.toolBar);
            this.toolBar$.complete();
            return this.toolBar;
        })));
    }
    /**
     * 销毁工具条
     * @return {?}
     */
    destroy() {
        zip(this.amaps.get(), this.get()).subscribe((/**
         * @param {?} __0
         * @return {?}
         */
        ([m, toolBar]) => {
            m.removeControl(toolBar);
            this.logger.d(TAG, 'toolBar destroyed.');
            this.toolBar = null;
        }));
    }
}
AmapToolBarService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
AmapToolBarService.ctorParameters = () => [
    { type: AMapService },
    { type: PluginLoaderService },
    { type: LoggerService },
    { type: NgZone }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    AmapToolBarService.prototype.toolBar;
    /**
     * @type {?}
     * @private
     */
    AmapToolBarService.prototype.toolBar$;
    /**
     * @type {?}
     * @private
     */
    AmapToolBarService.prototype.amaps;
    /**
     * @type {?}
     * @private
     */
    AmapToolBarService.prototype.plugin;
    /**
     * @type {?}
     * @private
     */
    AmapToolBarService.prototype.logger;
    /**
     * @type {?}
     * @private
     */
    AmapToolBarService.prototype.ngZone;
}
