// ngx-amap/esm5/directives/amap-heatmap/amap-heatmap.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: directives/amap-heatmap/amap-heatmap.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Injectable, NgZone } from '@angular/core';
import { ReplaySubject, zip } from 'rxjs';
import { map } from 'rxjs/operators';
import { AMapService } from '../../shared/amap.service';
import { PluginLoaderService } from '../../shared/plugin-loader.service';
import { LoggerService } from '../../shared/logger/logger.service';
/** @type {?} */
var TAG = 'AmapHeatmap';
export var AMapHeatmap;
(function (AMapHeatmap) {
    /**
     * @record
     */
    function Options() { }
    AMapHeatmap.Options = Options;
    if (false) {
        /**
         * 热力图中单个点的半径，默认：30，单位：pixel
         * @type {?|undefined}
         */
        Options.prototype.radius;
        /**
         * 热力图的渐变区间
         * @type {?|undefined}
         */
        Options.prototype.gradient;
        /**
         * 热力图透明度数组，取值范围[0,1]，0表示完全透明，1表示不透明
         * 默认：[0,1]
         * @type {?|undefined}
         */
        Options.prototype.opacity;
        /**
         * 支持的缩放级别范围，取值范围[3-18]
         * 默认：[3,18]
         * @type {?|undefined}
         */
        Options.prototype.zooms;
        /** @type {?|undefined} */
        Options.prototype.rejectMapMask;
        /** @type {?|undefined} */
        Options.prototype.visible;
        /** @type {?|undefined} */
        Options.prototype.radiusUnit;
        /** @type {?|undefined} */
        Options.prototype.blur;
        /** @type {?|undefined} */
        Options.prototype.zIndex;
        /** @type {?|undefined} */
        Options.prototype.renderOnZooming;
        /* Skipping unnamed member:
        ['3d']?: {
              heightScale?: number;
              heightBezier?: number[];
              gridSize?: number;
              drawGridLine?: boolean;
            };*/
    }
    /**
     * @record
     */
    function Data() { }
    AMapHeatmap.Data = Data;
    if (false) {
        /**
         * 经度
         * @type {?}
         */
        Data.prototype.lng;
        /**
         * 维度
         * @type {?}
         */
        Data.prototype.lat;
        /**
         * 权重
         * @type {?}
         */
        Data.prototype.count;
    }
    /**
     * @record
     */
    function DataSet() { }
    AMapHeatmap.DataSet = DataSet;
    if (false) {
        /**
         * 权重的最大值
         * @type {?|undefined}
         */
        DataSet.prototype.max;
        /**
         * 坐标数据集
         * @type {?}
         */
        DataSet.prototype.data;
    }
})(AMapHeatmap || (AMapHeatmap = {}));
var AmapHeatmapService = /** @class */ (function () {
    function AmapHeatmapService(amaps, plugin, logger, ngZone) {
        this.amaps = amaps;
        this.plugin = plugin;
        this.logger = logger;
        this.ngZone = ngZone;
        this.heatmap$ = new ReplaySubject(1);
    }
    /**
     * 获取热力图
     */
    /**
     * 获取热力图
     * @return {?}
     */
    AmapHeatmapService.prototype.get = /**
     * 获取热力图
     * @return {?}
     */
    function () {
        return this.heatmap$.asObservable();
    };
    /**
     * 创建热力图 AMap.Heatmap
     * @param options 选项
     */
    /**
     * 创建热力图 AMap.Heatmap
     * @param {?} options 选项
     * @return {?}
     */
    AmapHeatmapService.prototype.create = /**
     * 创建热力图 AMap.Heatmap
     * @param {?} options 选项
     * @return {?}
     */
    function (options) {
        var _this = this;
        return zip(this.amaps.get(), this.plugin.load('AMap.Heatmap')).pipe(map((/**
         * @param {?} __0
         * @return {?}
         */
        function (_a) {
            var _b = tslib_1.__read(_a, 1), m = _b[0];
            _this.heatmap = _this.ngZone.runOutsideAngular((/**
             * @return {?}
             */
            function () { return new AMap.Heatmap(m, options); }));
            _this.logger.d(TAG, 'new Heatmap created.');
            _this.heatmap$.next(_this.heatmap);
            _this.heatmap$.complete();
            return _this.heatmap;
        })));
    };
    /**
     * 销毁热力图
     */
    /**
     * 销毁热力图
     * @return {?}
     */
    AmapHeatmapService.prototype.destroy = /**
     * 销毁热力图
     * @return {?}
     */
    function () {
        var _this = this;
        this.get().subscribe((/**
         * @param {?} heatmap
         * @return {?}
         */
        function (heatmap) {
            heatmap.setMap(null);
            _this.logger.d(TAG, 'Heatmap destroyed.');
            _this.heatmap = null;
        }));
    };
    AmapHeatmapService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    AmapHeatmapService.ctorParameters = function () { return [
        { type: AMapService },
        { type: PluginLoaderService },
        { type: LoggerService },
        { type: NgZone }
    ]; };
    return AmapHeatmapService;
}());
export { AmapHeatmapService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    AmapHeatmapService.prototype.heatmap;
    /**
     * @type {?}
     * @private
     */
    AmapHeatmapService.prototype.heatmap$;
    /**
     * @type {?}
     * @private
     */
    AmapHeatmapService.prototype.amaps;
    /**
     * @type {?}
     * @private
     */
    AmapHeatmapService.prototype.plugin;
    /**
     * @type {?}
     * @private
     */
    AmapHeatmapService.prototype.logger;
    /**
     * @type {?}
     * @private
     */
    AmapHeatmapService.prototype.ngZone;
}
