// ngx-amap/esm2015/components/amap-text/amap-text.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: components/amap-text/amap-text.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, Output, EventEmitter, ContentChildren, QueryList, ElementRef, NgZone, } from '@angular/core';
import { zip } from 'rxjs';
import { AmapTextService } from './amap-text.service';
import { LoggerService } from '../../shared/logger/logger.service';
import { AMapOverlay, OverlayOptions } from '../../base/amap-overlay';
import { EventBinderService } from '../../shared/event-binder.service';
import { getOptions, ChangeFilter } from '../../utils';
import { PixelService } from '../../shared/pixel.service';
import { IconService } from '../../shared/icon.service';
import { AmapInfoWindowComponent } from '../amap-info-window/amap-info-window.component';
import { AMapService } from '../../shared/amap.service';
/** @type {?} */
const TAG = 'amap-text';
/** @type {?} */
const ALL_OPTIONS = [
    ...OverlayOptions,
    'text',
    'position',
    'anchor',
    'offset',
    'topWhenClick',
    'raiseOnDrag',
    'visible',
    'zIndex',
    'angle',
    'autoRotation',
    'animation',
    'shadow',
    'title',
    'style',
];
export class AmapTextComponent extends AMapOverlay {
    /**
     * @param {?} os
     * @param {?} binder
     * @param {?} amaps
     * @param {?} el
     * @param {?} pixels
     * @param {?} icons
     * @param {?} logger
     * @param {?} ngZone
     */
    constructor(os, binder, amaps, el, pixels, icons, logger, ngZone) {
        super(os, binder);
        this.os = os;
        this.binder = binder;
        this.amaps = amaps;
        this.el = el;
        this.pixels = pixels;
        this.icons = icons;
        this.logger = logger;
        this.ngZone = ngZone;
        /**
         * 额外: 是否隐藏
         */
        this.hidden = false;
        /**
         * 额外: 是否包含在点聚合中
         */
        this.inCluster = false;
        /**
         * 额外: 点击时是否显示信息窗体
         */
        this.openInfoWindow = true;
        // ---- Events ----
        this.naReady = new EventEmitter();
        // amap info window:
        this.infoWindowComponent = new QueryList();
        this.inited = false;
        /** @type {?} */
        const target = this.os.get();
        this.naMouseOut = this.binder.bindEvent(target, 'mouseout');
        this.naDragStart = this.binder.bindEvent(target, 'dragstart');
        this.naDragging = this.binder.bindEvent(target, 'dragging');
        this.naDragEnd = this.binder.bindEvent(target, 'dragend');
        this.naMoving = this.binder.bindEvent(target, 'moving');
        this.naMoveEnd = this.binder.bindEvent(target, 'moveend');
        this.naMoveAlong = this.binder.bindEvent(target, 'movealong');
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        if (this.subscription) {
            this.subscription.unsubscribe();
        }
        this.os.destroy();
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    ngOnChanges(changes) {
        /** @type {?} */
        const filter = ChangeFilter.of(changes);
        /** @type {?} */
        const marker = this.get();
        if (!this.inited) {
            // do not draw text marker when no poistion defined.
            if (!this.position) {
                return;
            }
            this.amaps.get().subscribe((/**
             * @return {?}
             */
            () => {
                this.logger.d(TAG, 'initializing ...');
                // bind info window events:
                this.subscription = this.binder.bindEvent(marker, 'click').subscribe((/**
                 * @return {?}
                 */
                () => {
                    if (this.openInfoWindow) {
                        this.infoWindowComponent.forEach((/**
                         * @param {?} w
                         * @return {?}
                         */
                        w => w.open()));
                    }
                }));
                /** @type {?} */
                const el = this.el.nativeElement.querySelector('.amap-text-content');
                this.text = this.text ? this.text : el.innerText;
                /** @type {?} */
                const options = getOptions(this, ALL_OPTIONS);
                if (this.shadow) {
                    options.shadow = this.icons.create(this.shadow);
                }
                if (this.offset) {
                    options.offset = this.pixels.create(this.offset);
                }
                this.logger.d(TAG, 'options:', options);
                this.os.create(options).subscribe((/**
                 * @param {?} m
                 * @return {?}
                 */
                m => {
                    this.ngZone.run((/**
                     * @return {?}
                     */
                    () => this.naReady.emit(m)));
                    this.logger.d(TAG, 'marker is ready.');
                }));
                this.inited = true;
                this.updateInfoWindow();
                this.updateInfoWindowPosition();
            }));
        }
        else {
            zip(filter.has('shadow'), marker).subscribe((/**
             * @param {?} __0
             * @return {?}
             */
            ([v, m]) => m.setShadow(this.icons.create(v))));
            zip(filter.has('offset'), marker).subscribe((/**
             * @param {?} __0
             * @return {?}
             */
            ([v, m]) => m.setOffset(this.pixels.create(v))));
            zip(filter.has('title'), marker).subscribe((/**
             * @param {?} __0
             * @return {?}
             */
            ([v, m]) => m.setTitle(v)));
            zip(filter.has('content'), marker).subscribe((/**
             * @param {?} __0
             * @return {?}
             */
            ([v, m]) => m.setContent(v)));
            zip(filter.has('extData'), marker).subscribe((/**
             * @param {?} __0
             * @return {?}
             */
            ([v, m]) => m.setExtData(v)));
            zip(filter.has('clickable'), marker).subscribe((/**
             * @param {?} __0
             * @return {?}
             */
            ([v, m]) => m.setClickable(!!v)));
            zip(filter.has('draggable'), marker).subscribe((/**
             * @param {?} __0
             * @return {?}
             */
            ([v, m]) => m.setDraggable(!!v)));
            zip(filter.has('visible'), marker).subscribe((/**
             * @param {?} __0
             * @return {?}
             */
            ([v, m]) => (v ? m.show() : m.hide())));
            zip(filter.has('cursor'), marker).subscribe((/**
             * @param {?} __0
             * @return {?}
             */
            ([v, m]) => m.setCursor(v)));
            zip(filter.has('animation'), marker).subscribe((/**
             * @param {?} __0
             * @return {?}
             */
            ([v, m]) => m.setAnimation(v)));
            zip(filter.has('angle'), marker).subscribe((/**
             * @param {?} __0
             * @return {?}
             */
            ([v, m]) => m.setAngle(v)));
            zip(filter.has('zIndex'), marker).subscribe((/**
             * @param {?} __0
             * @return {?}
             */
            ([v, m]) => m.setzIndex(v)));
            zip(filter.notEmpty('position'), marker).subscribe((/**
             * @param {?} __0
             * @return {?}
             */
            ([v, m]) => m.setPosition(v)));
            zip(filter.notEmpty('style'), marker).subscribe((/**
             * @param {?} __0
             * @return {?}
             */
            ([v, m]) => m.setStyle(v)));
        }
        zip(filter.has('isTop'), marker).subscribe((/**
         * @param {?} __0
         * @return {?}
         */
        ([v, m]) => m.setTop(!!v)));
        zip(filter.has('hidden'), marker).subscribe((/**
         * @param {?} __0
         * @return {?}
         */
        ([v, m]) => (v ? m.hide() : m.show())));
    }
    /**
     * @return {?}
     */
    ngAfterContentInit() {
        this.updateInfoWindow();
        this.infoWindowComponent.changes.subscribe((/**
         * @return {?}
         */
        () => this.updateInfoWindow()));
    }
    /**
     * @private
     * @return {?}
     */
    updateInfoWindow() {
        if (this.infoWindowComponent && this.inited) {
            if (this.infoWindowComponent.length > 1) {
                this.logger.e(TAG, 'Expected no more than 1 info window.');
                return;
            }
            /** @type {?} */
            const marker = this.os.get();
            this.infoWindowComponent.forEach((/**
             * @param {?} component
             * @return {?}
             */
            component => {
                component.hostMarker = marker;
            }));
        }
    }
    /**
     * @private
     * @return {?}
     */
    updateInfoWindowPosition() {
        if (this.infoWindowComponent && this.inited) {
            this.infoWindowComponent.forEach((/**
             * @param {?} component
             * @return {?}
             */
            component => {
                component.toggleOpen();
            }));
        }
    }
    /**
     * 获取已创建的 AMap.Text 对象
     * @return {?}
     */
    get() {
        return this.os.get();
    }
}
AmapTextComponent.decorators = [
    { type: Component, args: [{
                selector: 'amap-text',
                exportAs: 'text',
                template: "<div class='amap-text-inner'></div>\n<div class='amap-text-content'>\n  <ng-content></ng-content>\n</div>",
                providers: [AmapTextService]
            }] }
];
/** @nocollapse */
AmapTextComponent.ctorParameters = () => [
    { type: AmapTextService },
    { type: EventBinderService },
    { type: AMapService },
    { type: ElementRef },
    { type: PixelService },
    { type: IconService },
    { type: LoggerService },
    { type: NgZone }
];
AmapTextComponent.propDecorators = {
    text: [{ type: Input }],
    position: [{ type: Input }],
    anchor: [{ type: Input }],
    offset: [{ type: Input }],
    topWhenClick: [{ type: Input }],
    raiseOnDrag: [{ type: Input }],
    visible: [{ type: Input }],
    zIndex: [{ type: Input }],
    angle: [{ type: Input }],
    autoRotation: [{ type: Input }],
    animation: [{ type: Input }],
    shadow: [{ type: Input }],
    title: [{ type: Input }],
    style: [{ type: Input }],
    isTop: [{ type: Input }],
    hidden: [{ type: Input }],
    inCluster: [{ type: Input }],
    openInfoWindow: [{ type: Input }],
    naReady: [{ type: Output }],
    naMouseOut: [{ type: Output }],
    naDragStart: [{ type: Output }],
    naDragging: [{ type: Output }],
    naDragEnd: [{ type: Output }],
    naMoving: [{ type: Output }],
    naMoveEnd: [{ type: Output }],
    naMoveAlong: [{ type: Output }],
    infoWindowComponent: [{ type: ContentChildren, args: [AmapInfoWindowComponent,] }]
};
if (false) {
    /**
     * 点标记显示内容
     * @type {?}
     */
    AmapTextComponent.prototype.text;
    /**
     * 点标记在地图上显示的位置
     * @type {?}
     */
    AmapTextComponent.prototype.position;
    /**
     * 标记锚点
     * @type {?}
     */
    AmapTextComponent.prototype.anchor;
    /**
     * 点标记显示位置偏移量
     * @type {?}
     */
    AmapTextComponent.prototype.offset;
    /**
     * 鼠标点击时marker是否置顶
     * @type {?}
     */
    AmapTextComponent.prototype.topWhenClick;
    /**
     * 拖拽点标记时是否开启点标记离开地图的效果
     * @type {?}
     */
    AmapTextComponent.prototype.raiseOnDrag;
    /**
     * 点标记是否可见
     * @type {?}
     */
    AmapTextComponent.prototype.visible;
    /**
     * 点标记的叠加顺序
     * @type {?}
     */
    AmapTextComponent.prototype.zIndex;
    /**
     * 点标记的旋转角度
     * @type {?}
     */
    AmapTextComponent.prototype.angle;
    /**
     * 是否自动旋转
     * @type {?}
     */
    AmapTextComponent.prototype.autoRotation;
    /**
     * 点标记的动画效果
     * @type {?}
     */
    AmapTextComponent.prototype.animation;
    /**
     * 点标记阴影
     * @type {?}
     */
    AmapTextComponent.prototype.shadow;
    /**
     * 鼠标滑过点标记时的文字提示
     * @type {?}
     */
    AmapTextComponent.prototype.title;
    /**
     * 样式
     * @type {?}
     */
    AmapTextComponent.prototype.style;
    /**
     * 额外: 是否置顶
     * @type {?}
     */
    AmapTextComponent.prototype.isTop;
    /**
     * 额外: 是否隐藏
     * @type {?}
     */
    AmapTextComponent.prototype.hidden;
    /**
     * 额外: 是否包含在点聚合中
     * @type {?}
     */
    AmapTextComponent.prototype.inCluster;
    /**
     * 额外: 点击时是否显示信息窗体
     * @type {?}
     */
    AmapTextComponent.prototype.openInfoWindow;
    /** @type {?} */
    AmapTextComponent.prototype.naReady;
    /** @type {?} */
    AmapTextComponent.prototype.naMouseOut;
    /** @type {?} */
    AmapTextComponent.prototype.naDragStart;
    /** @type {?} */
    AmapTextComponent.prototype.naDragging;
    /** @type {?} */
    AmapTextComponent.prototype.naDragEnd;
    /** @type {?} */
    AmapTextComponent.prototype.naMoving;
    /** @type {?} */
    AmapTextComponent.prototype.naMoveEnd;
    /** @type {?} */
    AmapTextComponent.prototype.naMoveAlong;
    /** @type {?} */
    AmapTextComponent.prototype.infoWindowComponent;
    /**
     * @type {?}
     * @private
     */
    AmapTextComponent.prototype.inited;
    /**
     * @type {?}
     * @private
     */
    AmapTextComponent.prototype.subscription;
    /**
     * @type {?}
     * @protected
     */
    AmapTextComponent.prototype.os;
    /**
     * @type {?}
     * @protected
     */
    AmapTextComponent.prototype.binder;
    /**
     * @type {?}
     * @private
     */
    AmapTextComponent.prototype.amaps;
    /**
     * @type {?}
     * @private
     */
    AmapTextComponent.prototype.el;
    /**
     * @type {?}
     * @private
     */
    AmapTextComponent.prototype.pixels;
    /**
     * @type {?}
     * @private
     */
    AmapTextComponent.prototype.icons;
    /**
     * @type {?}
     * @private
     */
    AmapTextComponent.prototype.logger;
    /**
     * @type {?}
     * @private
     */
    AmapTextComponent.prototype.ngZone;
}
