// ngx-amap/esm2015/components/amap-text/amap-text.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: components/amap-text/amap-text.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, NgZone } from '@angular/core';
import { ReplaySubject } from 'rxjs';
import { map } from 'rxjs/operators';
import { AMapService } from '../../shared/amap.service';
import { LoggerService } from '../../shared/logger/logger.service';
/** @type {?} */
const TAG = 'AmapText';
export class AmapTextService {
    /**
     * @param {?} amaps
     * @param {?} logger
     * @param {?} ngZone
     */
    constructor(amaps, logger, ngZone) {
        this.amaps = amaps;
        this.logger = logger;
        this.ngZone = ngZone;
        this.text$ = new ReplaySubject(1);
    }
    /**
     * 获取点标记
     * @return {?}
     */
    get() {
        return this.text$.asObservable();
    }
    /**
     * 创建点标记 AMap.Text
     * @param {?} options 选项
     * @param {?=} addToMap 是否直接加进地图
     * @return {?}
     */
    create(options, addToMap = true) {
        return this.amaps.get().pipe(map((/**
         * @param {?} m
         * @return {?}
         */
        m => {
            if (addToMap) {
                options.map = m;
            }
            this.text = this.ngZone.runOutsideAngular((/**
             * @return {?}
             */
            () => new AMap.Text(options)));
            this.logger.d(TAG, 'new text created.');
            this.text$.next(this.text);
            this.text$.complete();
            return this.text;
        })));
    }
    /**
     * 销毁点标记
     * @return {?}
     */
    destroy() {
        this.get().subscribe((/**
         * @param {?} m
         * @return {?}
         */
        m => {
            m.setMap(null);
            this.logger.d(TAG, 'text destroyed.');
            this.text = null;
        }));
    }
}
AmapTextService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
AmapTextService.ctorParameters = () => [
    { type: AMapService },
    { type: LoggerService },
    { type: NgZone }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    AmapTextService.prototype.text;
    /**
     * @type {?}
     * @private
     */
    AmapTextService.prototype.text$;
    /**
     * @type {?}
     * @private
     */
    AmapTextService.prototype.amaps;
    /**
     * @type {?}
     * @private
     */
    AmapTextService.prototype.logger;
    /**
     * @type {?}
     * @private
     */
    AmapTextService.prototype.ngZone;
}
