// ngx-amap/esm5/shared/amap.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: shared/amap.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, NgZone } from '@angular/core';
import { ReplaySubject } from 'rxjs';
import { map } from 'rxjs/operators';
import { AMapLoaderService } from './amap-loader.service';
import { LoggerService } from './logger/logger.service';
/** @type {?} */
var TAG = 'AMap';
var AMapService = /** @class */ (function () {
    function AMapService(loader, logger, ngZone) {
        this.loader = loader;
        this.logger = logger;
        this.ngZone = ngZone;
        this.map$ = new ReplaySubject();
    }
    /**
     * 获取地图对象 AMap.Map
     */
    /**
     * 获取地图对象 AMap.Map
     * @return {?}
     */
    AMapService.prototype.get = /**
     * 获取地图对象 AMap.Map
     * @return {?}
     */
    function () {
        return this.map$.asObservable();
    };
    /**
     * 创建地图对象 AMap.Map
     * @param container 地图容器的DOM元素
     * @param options 选项
     */
    /**
     * 创建地图对象 AMap.Map
     * @param {?} container 地图容器的DOM元素
     * @param {?} options 选项
     * @return {?}
     */
    AMapService.prototype.create = /**
     * 创建地图对象 AMap.Map
     * @param {?} container 地图容器的DOM元素
     * @param {?} options 选项
     * @return {?}
     */
    function (container, options) {
        var _this = this;
        return this.loader.load().pipe(map((/**
         * @return {?}
         */
        function () {
            _this.map = _this.ngZone.runOutsideAngular((/**
             * @return {?}
             */
            function () { return new AMap.Map(container, options); }));
            _this.logger.d(TAG, 'new map created.');
            _this.map$.next(_this.map);
            _this.map$.complete();
            return _this.map;
        })));
    };
    /**
     * 销毁地图对象
     */
    /**
     * 销毁地图对象
     * @return {?}
     */
    AMapService.prototype.destroy = /**
     * 销毁地图对象
     * @return {?}
     */
    function () {
        var _this = this;
        this.get().subscribe((/**
         * @param {?} m
         * @return {?}
         */
        function (m) {
            m.destroy();
            _this.logger.d(TAG, 'map destroyed.');
            _this.map = null;
        }));
    };
    AMapService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    AMapService.ctorParameters = function () { return [
        { type: AMapLoaderService },
        { type: LoggerService },
        { type: NgZone }
    ]; };
    return AMapService;
}());
export { AMapService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    AMapService.prototype.map;
    /**
     * @type {?}
     * @private
     */
    AMapService.prototype.map$;
    /**
     * @type {?}
     * @private
     */
    AMapService.prototype.loader;
    /**
     * @type {?}
     * @private
     */
    AMapService.prototype.logger;
    /**
     * @type {?}
     * @private
     */
    AMapService.prototype.ngZone;
}
