// ngx-amap/esm5/shared/logger/debug-logger.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: shared/logger/debug-logger.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Injectable, Inject } from '@angular/core';
import { NGX_AMAP_CONFIG } from '../ngx-amap-config';
var DebugLoggerService = /** @class */ (function () {
    function DebugLoggerService(config) {
        if (config === void 0) { config = {}; }
        this.config = config;
        this.debugTags = [];
        this.debugAll = true;
        if (config.debugTags) {
            if (config.debugTags === '*') {
                this.debugAll = true;
            }
            else {
                this.debugAll = false;
                this.debugTags = config.debugTags.split(',');
            }
        }
    }
    /**
     * @param {...?} args
     * @return {?}
     */
    DebugLoggerService.prototype.i = /**
     * @param {...?} args
     * @return {?}
     */
    function () {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        console.log.apply(console, tslib_1.__spread(args));
    };
    /**
     * @param {...?} args
     * @return {?}
     */
    DebugLoggerService.prototype.e = /**
     * @param {...?} args
     * @return {?}
     */
    function () {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        console.error.apply(console, tslib_1.__spread(args));
    };
    /**
     * @param {...?} args
     * @return {?}
     */
    DebugLoggerService.prototype.w = /**
     * @param {...?} args
     * @return {?}
     */
    function () {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        console.warn.apply(console, tslib_1.__spread(args));
    };
    /**
     * @param {?} tag
     * @param {...?} args
     * @return {?}
     */
    DebugLoggerService.prototype.d = /**
     * @param {?} tag
     * @param {...?} args
     * @return {?}
     */
    function (tag) {
        var args = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            args[_i - 1] = arguments[_i];
        }
        if (this.debugAll || this.debugTags.includes(tag)) {
            console.log.apply(console, tslib_1.__spread(["[" + tag + "]"], args));
        }
    };
    DebugLoggerService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    DebugLoggerService.ctorParameters = function () { return [
        { type: undefined, decorators: [{ type: Inject, args: [NGX_AMAP_CONFIG,] }] }
    ]; };
    return DebugLoggerService;
}());
export { DebugLoggerService };
if (false) {
    /** @type {?} */
    DebugLoggerService.prototype.debugTags;
    /** @type {?} */
    DebugLoggerService.prototype.debugAll;
    /**
     * @type {?}
     * @private
     */
    DebugLoggerService.prototype.config;
}
