// ngx-amap/esm5/components/amap-text/amap-text.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: components/amap-text/amap-text.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, NgZone } from '@angular/core';
import { ReplaySubject } from 'rxjs';
import { map } from 'rxjs/operators';
import { AMapService } from '../../shared/amap.service';
import { LoggerService } from '../../shared/logger/logger.service';
/** @type {?} */
var TAG = 'AmapText';
var AmapTextService = /** @class */ (function () {
    function AmapTextService(amaps, logger, ngZone) {
        this.amaps = amaps;
        this.logger = logger;
        this.ngZone = ngZone;
        this.text$ = new ReplaySubject(1);
    }
    /**
     * 获取点标记
     */
    /**
     * 获取点标记
     * @return {?}
     */
    AmapTextService.prototype.get = /**
     * 获取点标记
     * @return {?}
     */
    function () {
        return this.text$.asObservable();
    };
    /**
     * 创建点标记 AMap.Text
     * @param options 选项
     * @param addToMap 是否直接加进地图
     */
    /**
     * 创建点标记 AMap.Text
     * @param {?} options 选项
     * @param {?=} addToMap 是否直接加进地图
     * @return {?}
     */
    AmapTextService.prototype.create = /**
     * 创建点标记 AMap.Text
     * @param {?} options 选项
     * @param {?=} addToMap 是否直接加进地图
     * @return {?}
     */
    function (options, addToMap) {
        var _this = this;
        if (addToMap === void 0) { addToMap = true; }
        return this.amaps.get().pipe(map((/**
         * @param {?} m
         * @return {?}
         */
        function (m) {
            if (addToMap) {
                options.map = m;
            }
            _this.text = _this.ngZone.runOutsideAngular((/**
             * @return {?}
             */
            function () { return new AMap.Text(options); }));
            _this.logger.d(TAG, 'new text created.');
            _this.text$.next(_this.text);
            _this.text$.complete();
            return _this.text;
        })));
    };
    /**
     * 销毁点标记
     */
    /**
     * 销毁点标记
     * @return {?}
     */
    AmapTextService.prototype.destroy = /**
     * 销毁点标记
     * @return {?}
     */
    function () {
        var _this = this;
        this.get().subscribe((/**
         * @param {?} m
         * @return {?}
         */
        function (m) {
            m.setMap(null);
            _this.logger.d(TAG, 'text destroyed.');
            _this.text = null;
        }));
    };
    AmapTextService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    AmapTextService.ctorParameters = function () { return [
        { type: AMapService },
        { type: LoggerService },
        { type: NgZone }
    ]; };
    return AmapTextService;
}());
export { AmapTextService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    AmapTextService.prototype.text;
    /**
     * @type {?}
     * @private
     */
    AmapTextService.prototype.text$;
    /**
     * @type {?}
     * @private
     */
    AmapTextService.prototype.amaps;
    /**
     * @type {?}
     * @private
     */
    AmapTextService.prototype.logger;
    /**
     * @type {?}
     * @private
     */
    AmapTextService.prototype.ngZone;
}
