// ngx-amap/esm2015/shared/logger/debug-logger.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: shared/logger/debug-logger.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, Inject } from '@angular/core';
import { NGX_AMAP_CONFIG } from '../ngx-amap-config';
export class DebugLoggerService {
    /**
     * @param {?=} config
     */
    constructor(config = {}) {
        this.config = config;
        this.debugTags = [];
        this.debugAll = true;
        if (config.debugTags) {
            if (config.debugTags === '*') {
                this.debugAll = true;
            }
            else {
                this.debugAll = false;
                this.debugTags = config.debugTags.split(',');
            }
        }
    }
    /**
     * @param {...?} args
     * @return {?}
     */
    i(...args) {
        console.log(...args);
    }
    /**
     * @param {...?} args
     * @return {?}
     */
    e(...args) {
        console.error(...args);
    }
    /**
     * @param {...?} args
     * @return {?}
     */
    w(...args) {
        console.warn(...args);
    }
    /**
     * @param {?} tag
     * @param {...?} args
     * @return {?}
     */
    d(tag, ...args) {
        if (this.debugAll || this.debugTags.includes(tag)) {
            console.log(`[${tag}]`, ...args);
        }
    }
}
DebugLoggerService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
DebugLoggerService.ctorParameters = () => [
    { type: undefined, decorators: [{ type: Inject, args: [NGX_AMAP_CONFIG,] }] }
];
if (false) {
    /** @type {?} */
    DebugLoggerService.prototype.debugTags;
    /** @type {?} */
    DebugLoggerService.prototype.debugAll;
    /**
     * @type {?}
     * @private
     */
    DebugLoggerService.prototype.config;
}
