// ngx-amap/esm2015/directives/amap-rectangle/amap-rectangle.directive.js
/**
 * @fileoverview added by tsickle
 * Generated from: directives/amap-rectangle/amap-rectangle.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Input, Directive, NgZone } from '@angular/core';
import { zip } from 'rxjs';
import { AMapPolygon, PolygonOptions } from '../../base/amap-polygon';
import { AmapRectangleService } from './amap-rectangle.service';
import { LoggerService } from '../../shared/logger/logger.service';
import { EventBinderService } from '../../shared/event-binder.service';
import { getOptions, ChangeFilter } from '../../utils';
/** @type {?} */
const TAG = 'amap-rectangle';
/** @type {?} */
const RectangleOptions = [...PolygonOptions, 'bounds'];
export class AmapRectangleDirective extends AMapPolygon {
    /**
     * @param {?} os
     * @param {?} binder
     * @param {?} logger
     * @param {?} ngZone
     */
    constructor(os, binder, logger, ngZone) {
        super(os, binder);
        this.os = os;
        this.binder = binder;
        this.logger = logger;
        this.ngZone = ngZone;
        this.inited = false;
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this.os.destroy();
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    ngOnChanges(changes) {
        /** @type {?} */
        const filter = ChangeFilter.of(changes);
        /** @type {?} */
        const rectangle = this.get();
        if (!this.inited) {
            this.logger.d(TAG, 'initializing ...');
            /** @type {?} */
            const options = this.options || getOptions(this, RectangleOptions);
            this.logger.d(TAG, 'options:', options);
            this.os.create(options).subscribe((/**
             * @param {?} m
             * @return {?}
             */
            m => {
                this.ngZone.run((/**
                 * @return {?}
                 */
                () => this.naReady.emit(m)));
                this.logger.d(TAG, 'rectangle is ready.');
            }));
            this.inited = true;
        }
        else {
            zip(filter.has('path'), rectangle).subscribe((/**
             * @param {?} __0
             * @return {?}
             */
            ([v, p]) => p.setPath(v)));
            zip(filter.has('options'), rectangle).subscribe((/**
             * @param {?} __0
             * @return {?}
             */
            ([v, p]) => p.setOptions(v || {})));
            zip(filter.has('bounds'), rectangle).subscribe((/**
             * @param {?} __0
             * @return {?}
             */
            ([v, p]) => p.setBounds(v)));
            zip(filter.has('extData'), rectangle).subscribe((/**
             * @param {?} __0
             * @return {?}
             */
            ([v, p]) => p.setExtData(v)));
        }
        zip(filter.has('hidden'), rectangle).subscribe((/**
         * @param {?} __0
         * @return {?}
         */
        ([v, p]) => (v ? p.hide() : p.show())));
        filter.has('editor').subscribe((/**
         * @param {?} v
         * @return {?}
         */
        v => this.os.toggleEditor(v)));
    }
    /**
     * 获取已创建的 AMap.Rectangle 对象
     * @return {?}
     */
    get() {
        return this.os.get();
    }
}
AmapRectangleDirective.decorators = [
    { type: Directive, args: [{
                selector: 'amap-rectangle',
                exportAs: 'rectangle',
                providers: [AmapRectangleService],
            },] }
];
/** @nocollapse */
AmapRectangleDirective.ctorParameters = () => [
    { type: AmapRectangleService },
    { type: EventBinderService },
    { type: LoggerService },
    { type: NgZone }
];
AmapRectangleDirective.propDecorators = {
    bounds: [{ type: Input }],
    options: [{ type: Input }]
};
if (false) {
    /**
     * 矩形的范围
     * @type {?}
     */
    AmapRectangleDirective.prototype.bounds;
    /**
     * 额外: 会覆盖其他属性的配置方式
     * @type {?}
     */
    AmapRectangleDirective.prototype.options;
    /**
     * @type {?}
     * @private
     */
    AmapRectangleDirective.prototype.inited;
    /**
     * @type {?}
     * @protected
     */
    AmapRectangleDirective.prototype.os;
    /**
     * @type {?}
     * @protected
     */
    AmapRectangleDirective.prototype.binder;
    /**
     * @type {?}
     * @private
     */
    AmapRectangleDirective.prototype.logger;
    /**
     * @type {?}
     * @private
     */
    AmapRectangleDirective.prototype.ngZone;
}
