// ngx-amap/esm2015/directives/amap-rectangle/amap-rectangle.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: directives/amap-rectangle/amap-rectangle.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, NgZone } from '@angular/core';
import { ReplaySubject, zip } from 'rxjs';
import { map, switchMapTo } from 'rxjs/operators';
import { AMapService } from '../../shared/amap.service';
import { LoggerService } from '../../shared/logger/logger.service';
import { PluginLoaderService } from '../../shared/plugin-loader.service';
/** @type {?} */
const TAG = 'AmapRectangle';
export class AmapRectangleService {
    /**
     * @param {?} amaps
     * @param {?} logger
     * @param {?} ngZone
     * @param {?} plugins
     */
    constructor(amaps, logger, ngZone, plugins) {
        this.amaps = amaps;
        this.logger = logger;
        this.ngZone = ngZone;
        this.plugins = plugins;
        this.rectangle$ = new ReplaySubject(1);
        this.editor$ = new ReplaySubject(1);
    }
    /**
     * 获取矩形
     * @return {?}
     */
    get() {
        return this.rectangle$.asObservable();
    }
    /**
     * 创建矩形 AMap.Rectangle
     * @param {?} options 选项
     * @return {?}
     */
    create(options) {
        return this.amaps.get().pipe(map((/**
         * @param {?} m
         * @return {?}
         */
        m => {
            options.map = m;
            this.rectangle = this.ngZone.runOutsideAngular((/**
             * @return {?}
             */
            () => new AMap.Rectangle(options)));
            this.logger.d(TAG, 'new Rectangle created.');
            this.rectangle$.next(this.rectangle);
            this.rectangle$.complete();
            return this.rectangle;
        })));
    }
    /**
     * 销毁矩形
     * @return {?}
     */
    destroy() {
        this.get().subscribe((/**
         * @param {?} m
         * @return {?}
         */
        m => {
            m.setMap(null);
            this.logger.d(TAG, 'Rectangle destroyed.');
            this.rectangle = null;
        }));
    }
    /**
     * @return {?}
     */
    getEditor() {
        // plugin is loading:
        return this.editor$.asObservable();
    }
    /**
     * 加载编辑插件
     * @return {?}
     */
    loadEditor() {
        return this.plugins.load('AMap.RectangleEditor');
    }
    /**
     * 创建编辑器
     * @return {?}
     */
    createEditor() {
        return this.loadEditor().pipe(switchMapTo(zip(this.amaps.get(), this.get())), map((/**
         * @param {?} __0
         * @return {?}
         */
        ([m, p]) => {
            this.editor = this.ngZone.runOutsideAngular((/**
             * @return {?}
             */
            () => new AMap.RectangleEditor(m, p)));
            this.logger.d(TAG, 'new RectangleEditor created.');
            this.editor$.next(this.editor);
            this.editor$.complete();
            return this.editor;
        })));
    }
    /**
     * 开关编辑器
     * @param {?} v 开关选项
     * @return {?}
     */
    toggleEditor(v) {
        if (v && !this.editor) {
            this.createEditor().subscribe((/**
             * @param {?} editor
             * @return {?}
             */
            editor => editor.open()));
            return;
        }
        if (this.editor) {
            if (v) {
                this.editor.open();
            }
            else {
                this.editor.close();
            }
        }
    }
}
AmapRectangleService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
AmapRectangleService.ctorParameters = () => [
    { type: AMapService },
    { type: LoggerService },
    { type: NgZone },
    { type: PluginLoaderService }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    AmapRectangleService.prototype.rectangle;
    /**
     * @type {?}
     * @private
     */
    AmapRectangleService.prototype.rectangle$;
    /**
     * @type {?}
     * @private
     */
    AmapRectangleService.prototype.editor;
    /**
     * @type {?}
     * @private
     */
    AmapRectangleService.prototype.editor$;
    /**
     * @type {?}
     * @private
     */
    AmapRectangleService.prototype.amaps;
    /**
     * @type {?}
     * @private
     */
    AmapRectangleService.prototype.logger;
    /**
     * @type {?}
     * @private
     */
    AmapRectangleService.prototype.ngZone;
    /**
     * @type {?}
     * @private
     */
    AmapRectangleService.prototype.plugins;
}
