// ngx-amap/esm2015/directives/amap-polygon/amap-polygon.directive.js
/**
 * @fileoverview added by tsickle
 * Generated from: directives/amap-polygon/amap-polygon.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, NgZone } from '@angular/core';
import { zip } from 'rxjs';
import { AMapPolygon, PolygonOptions } from '../../base/amap-polygon';
import { AmapPolygonService } from './amap-polygon.service';
import { LoggerService } from '../../shared/logger/logger.service';
import { EventBinderService } from '../../shared/event-binder.service';
import { getOptions, ChangeFilter } from '../../utils';
/** @type {?} */
const TAG = 'amap-polygon';
export class AmapPolygonDirective extends AMapPolygon {
    /**
     * @param {?} os
     * @param {?} binder
     * @param {?} logger
     * @param {?} ngZone
     */
    constructor(os, binder, logger, ngZone) {
        super(os, binder);
        this.os = os;
        this.binder = binder;
        this.logger = logger;
        this.ngZone = ngZone;
        this.inited = false;
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this.os.destroy();
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    ngOnChanges(changes) {
        /** @type {?} */
        const filter = ChangeFilter.of(changes);
        /** @type {?} */
        const polygon = this.get();
        if (!this.inited) {
            this.logger.d(TAG, 'initializing ...');
            /** @type {?} */
            const options = this.options || getOptions(this, PolygonOptions);
            this.logger.d(TAG, 'options:', options);
            this.os.create(options).subscribe((/**
             * @param {?} m
             * @return {?}
             */
            m => {
                this.ngZone.run((/**
                 * @return {?}
                 */
                () => this.naReady.emit(m)));
                this.logger.d(TAG, 'polygon is ready.');
            }));
            this.inited = true;
        }
        else {
            zip(filter.has('path'), polygon).subscribe((/**
             * @param {?} __0
             * @return {?}
             */
            ([v, p]) => p.setPath(v)));
            zip(filter.has('options'), polygon).subscribe((/**
             * @param {?} __0
             * @return {?}
             */
            ([v, p]) => p.setOptions(v || {})));
            zip(filter.has('extData'), polygon).subscribe((/**
             * @param {?} __0
             * @return {?}
             */
            ([v, p]) => p.setExtData(v)));
        }
        zip(filter.has('hidden'), polygon).subscribe((/**
         * @param {?} __0
         * @return {?}
         */
        ([v, p]) => (v ? p.hide() : p.show())));
        filter.has('editor').subscribe((/**
         * @param {?} v
         * @return {?}
         */
        v => this.os.toggleEditor(v)));
    }
    /**
     * 获取已创建的 AMap.Polygon 对象
     * @return {?}
     */
    get() {
        return this.os.get();
    }
}
AmapPolygonDirective.decorators = [
    { type: Directive, args: [{
                selector: 'amap-polygon',
                exportAs: 'polygon',
                providers: [AmapPolygonService],
            },] }
];
/** @nocollapse */
AmapPolygonDirective.ctorParameters = () => [
    { type: AmapPolygonService },
    { type: EventBinderService },
    { type: LoggerService },
    { type: NgZone }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    AmapPolygonDirective.prototype.inited;
    /**
     * @type {?}
     * @protected
     */
    AmapPolygonDirective.prototype.os;
    /**
     * @type {?}
     * @protected
     */
    AmapPolygonDirective.prototype.binder;
    /**
     * @type {?}
     * @private
     */
    AmapPolygonDirective.prototype.logger;
    /**
     * @type {?}
     * @private
     */
    AmapPolygonDirective.prototype.ngZone;
}
