// ngx-amap/esm2015/directives/amap-polygon/amap-polygon.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: directives/amap-polygon/amap-polygon.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, NgZone } from '@angular/core';
import { ReplaySubject, zip } from 'rxjs';
import { map, switchMapTo } from 'rxjs/operators';
import { AMapService } from '../../shared/amap.service';
import { LoggerService } from '../../shared/logger/logger.service';
import { PluginLoaderService } from '../../shared/plugin-loader.service';
/** @type {?} */
const TAG = 'AmapPolygon';
export class AmapPolygonService {
    /**
     * @param {?} amaps
     * @param {?} logger
     * @param {?} ngZone
     * @param {?} plugins
     */
    constructor(amaps, logger, ngZone, plugins) {
        this.amaps = amaps;
        this.logger = logger;
        this.ngZone = ngZone;
        this.plugins = plugins;
        this.polygon$ = new ReplaySubject(1);
        this.editor$ = new ReplaySubject(1);
    }
    /**
     * 获取折线
     * @return {?}
     */
    get() {
        return this.polygon$.asObservable();
    }
    /**
     * 创建折线 AMap.Polygon
     * @param {?} options 选项
     * @return {?}
     */
    create(options) {
        return this.amaps.get().pipe(map((/**
         * @param {?} m
         * @return {?}
         */
        m => {
            options.map = m;
            this.polygon = this.ngZone.runOutsideAngular((/**
             * @return {?}
             */
            () => new AMap.Polygon(options)));
            this.logger.d(TAG, 'new Polygon created.');
            this.polygon$.next(this.polygon);
            this.polygon$.complete();
            return this.polygon;
        })));
    }
    /**
     * 销毁折线
     * @return {?}
     */
    destroy() {
        this.get().subscribe((/**
         * @param {?} m
         * @return {?}
         */
        m => {
            m.setMap(null);
            this.logger.d(TAG, 'Polygon destroyed.');
            this.polygon = null;
        }));
    }
    /**
     * @return {?}
     */
    getEditor() {
        // plugin is loading:
        return this.editor$.asObservable();
    }
    /**
     * 加载编辑插件
     * @return {?}
     */
    loadEditor() {
        return this.plugins.load('AMap.PolyEditor');
    }
    /**
     * 创建编辑器
     * @return {?}
     */
    createEditor() {
        return this.loadEditor().pipe(switchMapTo(zip(this.amaps.get(), this.get())), map((/**
         * @param {?} __0
         * @return {?}
         */
        ([m, p]) => {
            this.editor = this.ngZone.runOutsideAngular((/**
             * @return {?}
             */
            () => new AMap.PolyEditor(m, p)));
            this.logger.d(TAG, 'new PolyEditor created.');
            this.editor$.next(this.editor);
            this.editor$.complete();
            return this.editor;
        })));
    }
    /**
     * 开关编辑器
     * @param {?} v 开关选项
     * @return {?}
     */
    toggleEditor(v) {
        if (v && !this.editor) {
            this.createEditor().subscribe((/**
             * @param {?} editor
             * @return {?}
             */
            editor => editor.open()));
            return;
        }
        if (this.editor) {
            if (v) {
                this.editor.open();
            }
            else {
                this.editor.close();
            }
        }
    }
}
AmapPolygonService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
AmapPolygonService.ctorParameters = () => [
    { type: AMapService },
    { type: LoggerService },
    { type: NgZone },
    { type: PluginLoaderService }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    AmapPolygonService.prototype.polygon;
    /**
     * @type {?}
     * @private
     */
    AmapPolygonService.prototype.polygon$;
    /**
     * @type {?}
     * @private
     */
    AmapPolygonService.prototype.editor;
    /**
     * @type {?}
     * @private
     */
    AmapPolygonService.prototype.editor$;
    /**
     * @type {?}
     * @private
     */
    AmapPolygonService.prototype.amaps;
    /**
     * @type {?}
     * @private
     */
    AmapPolygonService.prototype.logger;
    /**
     * @type {?}
     * @private
     */
    AmapPolygonService.prototype.ngZone;
    /**
     * @type {?}
     * @private
     */
    AmapPolygonService.prototype.plugins;
}
