// ngx-amap/esm5/directives/amap-circle-marker/amap-circle-marker.directive.js
/**
 * @fileoverview added by tsickle
 * Generated from: directives/amap-circle-marker/amap-circle-marker.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Directive, NgZone, } from '@angular/core';
import { zip } from 'rxjs';
import { AMapCircle, CircleOptions } from '../../base/amap-circle';
import { AmapCircleMarkerService } from './amap-circle-marker.service';
import { LoggerService } from '../../shared/logger/logger.service';
import { EventBinderService } from '../../shared/event-binder.service';
import { getOptions, ChangeFilter } from '../../utils';
/** @type {?} */
var TAG = 'amap-circle-marker';
var AmapCircleMarkerDirective = /** @class */ (function (_super) {
    tslib_1.__extends(AmapCircleMarkerDirective, _super);
    function AmapCircleMarkerDirective(os, binder, logger, ngZone) {
        var _this = _super.call(this, os, binder) || this;
        _this.os = os;
        _this.binder = binder;
        _this.logger = logger;
        _this.ngZone = ngZone;
        _this.inited = false;
        return _this;
    }
    /**
     * @return {?}
     */
    AmapCircleMarkerDirective.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this.os.destroy();
    };
    /**
     * @param {?} changes
     * @return {?}
     */
    AmapCircleMarkerDirective.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        var _this = this;
        /** @type {?} */
        var filter = ChangeFilter.of(changes);
        /** @type {?} */
        var circle = this.get();
        if (!this.inited) {
            this.logger.d(TAG, 'initializing ...');
            /** @type {?} */
            var options = this.options || getOptions(this, CircleOptions);
            this.logger.d(TAG, 'options:', options);
            this.os.create(options).subscribe((/**
             * @param {?} m
             * @return {?}
             */
            function (m) {
                _this.ngZone.run((/**
                 * @return {?}
                 */
                function () { return _this.naReady.emit(m); }));
                _this.logger.d(TAG, 'circle marker is ready.');
            }));
            this.inited = true;
        }
        else {
            zip(filter.has('center'), circle).subscribe((/**
             * @param {?} __0
             * @return {?}
             */
            function (_a) {
                var _b = tslib_1.__read(_a, 2), v = _b[0], p = _b[1];
                return p.setCenter(v);
            }));
            zip(filter.has('options'), circle).subscribe((/**
             * @param {?} __0
             * @return {?}
             */
            function (_a) {
                var _b = tslib_1.__read(_a, 2), v = _b[0], p = _b[1];
                return p.setOptions(v || {});
            }));
            zip(filter.has('radius'), circle).subscribe((/**
             * @param {?} __0
             * @return {?}
             */
            function (_a) {
                var _b = tslib_1.__read(_a, 2), v = _b[0], p = _b[1];
                return p.setRadius(v);
            }));
            zip(filter.has('extData'), circle).subscribe((/**
             * @param {?} __0
             * @return {?}
             */
            function (_a) {
                var _b = tslib_1.__read(_a, 2), v = _b[0], p = _b[1];
                return p.setExtData(v);
            }));
        }
        zip(filter.has('hidden'), circle).subscribe((/**
         * @param {?} __0
         * @return {?}
         */
        function (_a) {
            var _b = tslib_1.__read(_a, 2), v = _b[0], p = _b[1];
            return (v ? p.hide() : p.show());
        }));
    };
    /**
     * 获取已创建的 AMap.CircleMarker 对象
     */
    /**
     * 获取已创建的 AMap.CircleMarker 对象
     * @return {?}
     */
    AmapCircleMarkerDirective.prototype.get = /**
     * 获取已创建的 AMap.CircleMarker 对象
     * @return {?}
     */
    function () {
        return this.os.get();
    };
    AmapCircleMarkerDirective.decorators = [
        { type: Directive, args: [{
                    selector: 'amap-circle-marker',
                    exportAs: 'circle-marker',
                    providers: [AmapCircleMarkerService],
                },] }
    ];
    /** @nocollapse */
    AmapCircleMarkerDirective.ctorParameters = function () { return [
        { type: AmapCircleMarkerService },
        { type: EventBinderService },
        { type: LoggerService },
        { type: NgZone }
    ]; };
    return AmapCircleMarkerDirective;
}(AMapCircle));
export { AmapCircleMarkerDirective };
if (false) {
    /**
     * @type {?}
     * @private
     */
    AmapCircleMarkerDirective.prototype.inited;
    /**
     * @type {?}
     * @protected
     */
    AmapCircleMarkerDirective.prototype.os;
    /**
     * @type {?}
     * @protected
     */
    AmapCircleMarkerDirective.prototype.binder;
    /**
     * @type {?}
     * @private
     */
    AmapCircleMarkerDirective.prototype.logger;
    /**
     * @type {?}
     * @private
     */
    AmapCircleMarkerDirective.prototype.ngZone;
}
