// ngx-amap/esm5/directives/amap-circle-marker/amap-circle-marker.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: directives/amap-circle-marker/amap-circle-marker.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, NgZone } from '@angular/core';
import { ReplaySubject } from 'rxjs';
import { map } from 'rxjs/operators';
import { AMapService } from '../../shared/amap.service';
import { LoggerService } from '../../shared/logger/logger.service';
/** @type {?} */
var TAG = 'AmapCircleMarker';
var AmapCircleMarkerService = /** @class */ (function () {
    function AmapCircleMarkerService(amaps, logger, ngZone) {
        this.amaps = amaps;
        this.logger = logger;
        this.ngZone = ngZone;
        this.circle$ = new ReplaySubject(1);
    }
    /**
     * 获取圆点标记
     */
    /**
     * 获取圆点标记
     * @return {?}
     */
    AmapCircleMarkerService.prototype.get = /**
     * 获取圆点标记
     * @return {?}
     */
    function () {
        return this.circle$.asObservable();
    };
    /**
     * 创建圆点标记 AMap.CircleMarker
     * @param options 选项
     */
    /**
     * 创建圆点标记 AMap.CircleMarker
     * @param {?} options 选项
     * @return {?}
     */
    AmapCircleMarkerService.prototype.create = /**
     * 创建圆点标记 AMap.CircleMarker
     * @param {?} options 选项
     * @return {?}
     */
    function (options) {
        var _this = this;
        return this.amaps.get().pipe(map((/**
         * @param {?} m
         * @return {?}
         */
        function (m) {
            options.map = m;
            _this.circle = _this.ngZone.runOutsideAngular((/**
             * @return {?}
             */
            function () { return new AMap.CircleMarker(options); }));
            _this.logger.d(TAG, 'new CircleMarker created.');
            _this.circle$.next(_this.circle);
            _this.circle$.complete();
            return _this.circle;
        })));
    };
    /**
     * 销毁圆点标记
     */
    /**
     * 销毁圆点标记
     * @return {?}
     */
    AmapCircleMarkerService.prototype.destroy = /**
     * 销毁圆点标记
     * @return {?}
     */
    function () {
        var _this = this;
        this.get().subscribe((/**
         * @param {?} m
         * @return {?}
         */
        function (m) {
            m.setMap(null);
            _this.logger.d(TAG, 'CircleMarker destroyed.');
            _this.circle = null;
        }));
    };
    AmapCircleMarkerService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    AmapCircleMarkerService.ctorParameters = function () { return [
        { type: AMapService },
        { type: LoggerService },
        { type: NgZone }
    ]; };
    return AmapCircleMarkerService;
}());
export { AmapCircleMarkerService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    AmapCircleMarkerService.prototype.circle;
    /**
     * @type {?}
     * @private
     */
    AmapCircleMarkerService.prototype.circle$;
    /**
     * @type {?}
     * @private
     */
    AmapCircleMarkerService.prototype.amaps;
    /**
     * @type {?}
     * @private
     */
    AmapCircleMarkerService.prototype.logger;
    /**
     * @type {?}
     * @private
     */
    AmapCircleMarkerService.prototype.ngZone;
}
