// ngx-amap/esm2015/directives/ui-simple-marker/ui-simple-marker.directive.js
/**
 * @fileoverview added by tsickle
 * Generated from: directives/ui-simple-marker/ui-simple-marker.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, NgZone } from '@angular/core';
import { UISimpleMarkerService } from './ui-simple-marker.service';
import { UISimpleMarker } from './ui-simple-marker';
import { EventBinderService } from '../../shared/event-binder.service';
import { AMapService } from '../../shared/amap.service';
import { PixelService } from '../../shared/pixel.service';
import { IconService } from '../../shared/icon.service';
import { MarkerLabelService } from '../../shared/marker-label.service';
import { LoggerService } from '../../shared/logger/logger.service';
export class UISimpleMarkerDirective extends UISimpleMarker {
    /**
     * @param {?} os
     * @param {?} binder
     * @param {?} amaps
     * @param {?} pixels
     * @param {?} icons
     * @param {?} mlabels
     * @param {?} logger
     * @param {?} ngZone
     */
    constructor(os, binder, amaps, pixels, icons, mlabels, logger, ngZone) {
        super(os, binder, amaps, pixels, icons, mlabels, logger, ngZone);
        this.os = os;
        this.binder = binder;
        this.amaps = amaps;
        this.pixels = pixels;
        this.icons = icons;
        this.mlabels = mlabels;
        this.logger = logger;
        this.ngZone = ngZone;
    }
}
UISimpleMarkerDirective.decorators = [
    { type: Directive, args: [{
                selector: 'ui-simple-marker',
                exportAs: 'simpleMarker',
                providers: [UISimpleMarkerService],
            },] }
];
/** @nocollapse */
UISimpleMarkerDirective.ctorParameters = () => [
    { type: UISimpleMarkerService },
    { type: EventBinderService },
    { type: AMapService },
    { type: PixelService },
    { type: IconService },
    { type: MarkerLabelService },
    { type: LoggerService },
    { type: NgZone }
];
if (false) {
    /**
     * @type {?}
     * @protected
     */
    UISimpleMarkerDirective.prototype.os;
    /**
     * @type {?}
     * @protected
     */
    UISimpleMarkerDirective.prototype.binder;
    /**
     * @type {?}
     * @protected
     */
    UISimpleMarkerDirective.prototype.amaps;
    /**
     * @type {?}
     * @protected
     */
    UISimpleMarkerDirective.prototype.pixels;
    /**
     * @type {?}
     * @protected
     */
    UISimpleMarkerDirective.prototype.icons;
    /**
     * @type {?}
     * @protected
     */
    UISimpleMarkerDirective.prototype.mlabels;
    /**
     * @type {?}
     * @protected
     */
    UISimpleMarkerDirective.prototype.logger;
    /**
     * @type {?}
     * @protected
     */
    UISimpleMarkerDirective.prototype.ngZone;
}
